// @cloudflare/sandbox/dist/dist-CmfvOT-w.js
//#region ../shared/dist/env.js
/**
* Safely extract a string value from an environment object
*
* @param env - Environment object with dynamic keys
* @param key - The environment variable key to access
* @returns The string value if present and is a string, undefined otherwise
*/
function getEnvString(env, key) {
	const value = env?.[key];
	return typeof value === "string" ? value : void 0;
}
/**
* Filter environment variables object to only include string values.
* Skips undefined, null, and non-string values.
*
* Use this when you only need the defined values (e.g., for per-command env
* where undefined means "don't override").
*
* @param envVars - Object that may contain undefined values
* @returns Clean object with only string values
*/
function filterEnvVars(envVars) {
	const filtered = {};
	for (const [key, value] of Object.entries(envVars)) if (value != null && typeof value === "string") filtered[key] = value;
	return filtered;
}
/**
* Partition environment variables into values to set and keys to unset.
*
* - String values → toSet (will be exported)
* - undefined/null → toUnset (will be unset)
*
* This enables idiomatic JS patterns where undefined means "remove":
* ```typescript
* await sandbox.setEnvVars({
*   API_KEY: 'new-key',        // will be set
*   OLD_VAR: undefined,        // will be unset
* });
* ```
*/
function partitionEnvVars(envVars) {
	const toSet = {};
	const toUnset = [];
	for (const [key, value] of Object.entries(envVars)) if (value != null && typeof value === "string") toSet[key] = value;
	else toUnset.push(key);
	return {
		toSet,
		toUnset
	};
}

//#endregion
//#region ../shared/dist/logger/sanitize.js
/**
* Log-only sanitization helpers
*
* These functions redact sensitive values for logging output.
* They MUST NOT be used to mutate command strings before execution.
*/
/**
* Sensitive query parameter names to redact from URLs.
* Anchored to query string context ([?&]) to avoid matching path segments.
* Value matching stops at & and common URL/command delimiters.
*/
const SENSITIVE_PARAMS = /([?&])(X-Amz-Credential|X-Amz-Signature|X-Amz-Security-Token|token|secret|password)=[^&\s"'`<>]*/gi;
/**
* Redact credentials from URLs for secure logging
*
* Replaces any credentials (username:password, tokens, etc.) embedded
* in URLs with ****** to prevent sensitive data exposure in logs.
* Works with URLs embedded in text.
*
* @param text - String that may contain URLs with credentials
* @returns String with credentials redacted from any URLs
*/
function redactCredentials(text) {
	let result = text;
	let pos = 0;
	while (pos < result.length) {
		const httpPos = result.indexOf("http://", pos);
		const httpsPos = result.indexOf("https://", pos);
		let protocolPos = -1;
		let protocolLen = 0;
		if (httpPos === -1 && httpsPos === -1) break;
		if (httpPos !== -1 && (httpsPos === -1 || httpPos < httpsPos)) {
			protocolPos = httpPos;
			protocolLen = 7;
		} else {
			protocolPos = httpsPos;
			protocolLen = 8;
		}
		const searchStart = protocolPos + protocolLen;
		const atPos = result.indexOf("@", searchStart);
		let urlEnd = searchStart;
		while (urlEnd < result.length) {
			const char = result[urlEnd];
			if (/[\s"'`<>,;{}[\]]/.test(char)) break;
			urlEnd++;
		}
		if (atPos !== -1 && atPos < urlEnd) {
			result = `${result.substring(0, searchStart)}******${result.substring(atPos)}`;
			pos = searchStart + 6;
		} else pos = protocolPos + protocolLen;
	}
	return result;
}
/**
* Redact sensitive query parameters from URLs
*
* Strips X-Amz-Credential, X-Amz-Signature, X-Amz-Security-Token,
* token, secret, and password query params from URLs. Returns
* non-URL strings unchanged.
*
* @param input - String that may contain URLs with sensitive params
* @returns String with sensitive params replaced by REDACTED
*/
function redactSensitiveParams(input) {
	if (!input.includes("?") || !input.includes("=")) return input;
	return input.replace(SENSITIVE_PARAMS, "$1$2=REDACTED");
}
/**
* Redact sensitive data from a command string for logging
*
* Composes redactCredentials (URL credentials) and redactSensitiveParams
* (presigned URL query params). For log values only — never mutate
* command strings before execution.
*
* @param command - Command string to sanitize for logging
* @returns Sanitized command string
*/
function redactCommand(command) {
	return redactSensitiveParams(redactCredentials(command));
}
/**
* Truncate a string for log output with a truncation indicator
*
* @param value - String to potentially truncate
* @param maxLen - Maximum length before truncation (default 120)
* @returns Object with truncated value and boolean flag
*/
function truncateForLog(value, maxLen = 120) {
	if (value.length <= maxLen) return {
		value,
		truncated: false
	};
	const cutoff = Math.max(0, maxLen - 3);
	return {
		value: `${value.substring(0, cutoff)}...`,
		truncated: true
	};
}

//#endregion
//#region ../shared/dist/git.js
/**
* Fallback repository name used when URL parsing fails
*/
const FALLBACK_REPO_NAME = "repository";
/**
* Extract repository name from a Git URL
*
* Supports multiple URL formats:
* - HTTPS: https://github.com/user/repo.git → repo
* - HTTPS without .git: https://github.com/user/repo → repo
* - SSH: git@github.com:user/repo.git → repo
* - GitLab/others: https://gitlab.com/org/project.git → project
*
* @param repoUrl - Git repository URL (HTTPS or SSH format)
* @returns Repository name extracted from URL, or 'repository' as fallback
*/
function extractRepoName(repoUrl) {
	try {
		const pathParts = new URL(repoUrl).pathname.split("/");
		const lastPart = pathParts[pathParts.length - 1];
		if (lastPart) return lastPart.replace(/\.git$/, "");
	} catch {}
	if (repoUrl.includes(":") || repoUrl.includes("/")) {
		const segments = repoUrl.split(/[:/]/).filter(Boolean);
		const lastSegment = segments[segments.length - 1];
		if (lastSegment) return lastSegment.replace(/\.git$/, "");
	}
	return FALLBACK_REPO_NAME;
}
/**
* Sanitize data by redacting credentials from any strings
* Recursively processes objects and arrays to ensure credentials are never leaked
*/
function sanitizeGitData(data) {
	if (typeof data === "string") return redactCommand(data);
	if (data === null || data === void 0) return data;
	if (Array.isArray(data)) return data.map((item) => sanitizeGitData(item));
	if (typeof data === "object") {
		const result = {};
		for (const [key, value] of Object.entries(data)) result[key] = sanitizeGitData(value);
		return result;
	}
	return data;
}
/**
* Logger wrapper that automatically sanitizes git credentials
*/
var GitLogger = class GitLogger {
	baseLogger;
	constructor(baseLogger) {
		this.baseLogger = baseLogger;
	}
	sanitizeContext(context) {
		return context ? sanitizeGitData(context) : context;
	}
	sanitizeError(error) {
		if (!error) return error;
		const sanitized = new Error(redactCommand(error.message));
		sanitized.name = error.name;
		if (error.stack) sanitized.stack = redactCommand(error.stack);
		const sanitizedRecord = sanitized;
		const errorRecord = error;
		for (const key of Object.keys(error)) if (key !== "message" && key !== "stack" && key !== "name") sanitizedRecord[key] = sanitizeGitData(errorRecord[key]);
		return sanitized;
	}
	debug(message, context) {
		this.baseLogger.debug(message, this.sanitizeContext(context));
	}
	info(message, context) {
		this.baseLogger.info(message, this.sanitizeContext(context));
	}
	warn(message, context) {
		this.baseLogger.warn(message, this.sanitizeContext(context));
	}
	error(message, error, context) {
		this.baseLogger.error(message, this.sanitizeError(error), this.sanitizeContext(context));
	}
	child(context) {
		const sanitized = sanitizeGitData(context);
		return new GitLogger(this.baseLogger.child(sanitized));
	}
};

//#endregion
//#region ../shared/dist/interpreter-types.js
var Execution = class {
	code;
	context;
	/**
	* All results from the execution
	*/
	results = [];
	/**
	* Accumulated stdout and stderr
	*/
	logs = {
		stdout: [],
		stderr: []
	};
	/**
	* Execution error if any
	*/
	error;
	/**
	* Execution count (for interpreter)
	*/
	executionCount;
	constructor(code, context) {
		this.code = code;
		this.context = context;
	}
	/**
	* Convert to a plain object for serialization
	*/
	toJSON() {
		return {
			code: this.code,
			logs: this.logs,
			error: this.error,
			executionCount: this.executionCount,
			results: this.results.map((result) => ({
				text: result.text,
				html: result.html,
				png: result.png,
				jpeg: result.jpeg,
				svg: result.svg,
				latex: result.latex,
				markdown: result.markdown,
				javascript: result.javascript,
				json: result.json,
				chart: result.chart,
				data: result.data
			}))
		};
	}
};
var ResultImpl = class {
	raw;
	constructor(raw) {
		this.raw = raw;
	}
	get text() {
		return this.raw.text || this.raw.data?.["text/plain"];
	}
	get html() {
		return this.raw.html || this.raw.data?.["text/html"];
	}
	get png() {
		return this.raw.png || this.raw.data?.["image/png"];
	}
	get jpeg() {
		return this.raw.jpeg || this.raw.data?.["image/jpeg"];
	}
	get svg() {
		return this.raw.svg || this.raw.data?.["image/svg+xml"];
	}
	get latex() {
		return this.raw.latex || this.raw.data?.["text/latex"];
	}
	get markdown() {
		return this.raw.markdown || this.raw.data?.["text/markdown"];
	}
	get javascript() {
		return this.raw.javascript || this.raw.data?.["application/javascript"];
	}
	get json() {
		return this.raw.json || this.raw.data?.["application/json"];
	}
	get chart() {
		return this.raw.chart;
	}
	get data() {
		return this.raw.data;
	}
	formats() {
		const formats = [];
		if (this.text) formats.push("text");
		if (this.html) formats.push("html");
		if (this.png) formats.push("png");
		if (this.jpeg) formats.push("jpeg");
		if (this.svg) formats.push("svg");
		if (this.latex) formats.push("latex");
		if (this.markdown) formats.push("markdown");
		if (this.javascript) formats.push("javascript");
		if (this.json) formats.push("json");
		if (this.chart) formats.push("chart");
		return formats;
	}
};

//#endregion
//#region ../shared/dist/logger/canonical.js
/** Events that are low-value at info on success */
const DEBUG_ON_SUCCESS = new Set([
	"session.create",
	"session.destroy",
	"file.read",
	"file.write",
	"file.delete",
	"file.mkdir"
]);
function resolveLogLevel(payload, options) {
	if (payload.outcome === "error") return "error";
	if (options?.successLevel) return options.successLevel;
	if (payload.origin === "internal") return "debug";
	if (DEBUG_ON_SUCCESS.has(payload.event)) return "debug";
	return "info";
}
/**
* Sanitize an Error object by redacting sensitive data from message and stack.
* Produces a copy so the caller's original Error is not mutated.
*/
function sanitizeError(error) {
	if (!error) return void 0;
	const sanitized = new Error(redactCommand(error.message));
	sanitized.name = error.name;
	sanitized.stack = error.stack ? redactCommand(error.stack) : void 0;
	return sanitized;
}
/**
* Sanitize and prepare payload fields for both message building and context emission.
* Called once by logCanonicalEvent to avoid double-redaction.
*/
function sanitizePayload(payload) {
	if (payload.command === void 0) return { commandTruncated: false };
	const { value, truncated } = truncateForLog(redactCommand(payload.command));
	return {
		sanitizedCommand: value,
		commandTruncated: truncated
	};
}
/**
* Build a human-readable canonical event message for dashboards and log viewers.
*
* Format: `{event} {outcome} {key_context} [— {reason}] ({durationMs}ms[, {sizeBytes}B])`
*
* The if/else chain for key context has implicit priority: command > path >
* sessionId > port > repoUrl > pid. If a payload has multiple, only the
* highest-priority one appears in the message. All fields are still present
* as discrete queryable keys in the structured log context.
*/
function buildMessage(payload, sanitizedCommand) {
	const { event } = payload;
	if (event === "version.check") {
		const parts$1 = ["version.check"];
		if (payload.sdkVersion) parts$1.push(`sdk=${payload.sdkVersion}`);
		if (payload.containerVersion) parts$1.push(`container=${payload.containerVersion}`);
		if (payload.versionOutcome && payload.versionOutcome !== "compatible") parts$1.push(`(${payload.versionOutcome})`);
		return parts$1.join(" ");
	}
	const parts = [event, payload.outcome];
	if (sanitizedCommand !== void 0) parts.push(sanitizedCommand);
	else if (payload.command !== void 0) {
		const { value } = truncateForLog(redactCommand(payload.command));
		parts.push(value);
	} else if (payload.path !== void 0) parts.push(payload.path);
	else if (event.includes("session") && payload.sessionId !== void 0) parts.push(payload.sessionId);
	else if (payload.port !== void 0) parts.push(String(payload.port));
	else if (payload.repoUrl !== void 0) {
		let gitContext = payload.repoUrl;
		if (payload.branch !== void 0) gitContext += ` ${payload.branch}`;
		parts.push(gitContext);
	} else if (payload.pid !== void 0) parts.push(String(payload.pid));
	else if (payload.backupId !== void 0) parts.push(payload.backupId);
	else if (payload.repoPath !== void 0) {
		let gitContext = payload.repoPath;
		if (payload.branch !== void 0) gitContext += ` branch=${payload.branch}`;
		parts.push(gitContext);
	} else if (payload.mountsProcessed !== void 0) {
		let destroyContext = `${payload.mountsProcessed} mounts`;
		if (payload.mountFailures) destroyContext += `, ${payload.mountFailures} failed`;
		parts.push(destroyContext);
	} else if (payload.mountPath !== void 0) parts.push(payload.mountPath);
	if (payload.outcome === "error") {
		if (payload.errorMessage !== void 0) parts.push(`\u2014 ${payload.errorMessage}`);
		else if (payload.exitCode !== void 0) parts.push(`\u2014 exitCode=${payload.exitCode}`);
	}
	const durationSuffix = payload.sizeBytes !== void 0 ? `(${payload.durationMs}ms, ${payload.sizeBytes}B)` : `(${payload.durationMs}ms)`;
	parts.push(durationSuffix);
	return parts.join(" ");
}
/**
* Log a canonical event — the single entry point for all structured operational events.
*
* Sanitizes command fields once, builds the message, selects log level from
* outcome, and emits a structured log entry with the full payload as context.
*/
function logCanonicalEvent(logger, payload, options) {
	const resolvedErrorMessage = payload.errorMessage ?? payload.error?.message;
	const sanitizedErrorMessage = resolvedErrorMessage ? redactCommand(resolvedErrorMessage) : void 0;
	const enrichedPayload = sanitizedErrorMessage !== void 0 ? {
		...payload,
		errorMessage: sanitizedErrorMessage
	} : payload;
	const { sanitizedCommand, commandTruncated } = sanitizePayload(enrichedPayload);
	const message = buildMessage(enrichedPayload, sanitizedCommand);
	const context = {};
	for (const [key, value] of Object.entries(enrichedPayload)) {
		if (key === "error") continue;
		context[key] = value;
	}
	if (sanitizedCommand !== void 0) {
		context.command = sanitizedCommand;
		if (commandTruncated) context.commandTruncated = true;
	}
	const level = resolveLogLevel(enrichedPayload, options);
	if (level === "error") logger.error(message, sanitizeError(payload.error), context);
	else if (level === "warn") logger.warn(message, context);
	else if (level === "debug") logger.debug(message, context);
	else logger.info(message, context);
}

//#endregion
//#region ../shared/dist/logger/types.js
/**
* Logger types for Cloudflare Sandbox SDK
*
* Provides structured, trace-aware logging across Worker, Durable Object, and Container.
*/
/**
* Log levels (from most to least verbose)
*/
var LogLevel;
(function(LogLevel$1) {
	LogLevel$1[LogLevel$1["DEBUG"] = 0] = "DEBUG";
	LogLevel$1[LogLevel$1["INFO"] = 1] = "INFO";
	LogLevel$1[LogLevel$1["WARN"] = 2] = "WARN";
	LogLevel$1[LogLevel$1["ERROR"] = 3] = "ERROR";
})(LogLevel || (LogLevel = {}));

//#endregion
//#region ../shared/dist/logger/logger.js
/**
* Logger implementation
*/
/**
* ANSI color codes for terminal output
*/
const COLORS = {
	reset: "\x1B[0m",
	debug: "\x1B[36m",
	info: "\x1B[32m",
	warn: "\x1B[33m",
	error: "\x1B[31m",
	dim: "\x1B[2m"
};
/**
* CloudflareLogger implements structured logging with support for
* both JSON output (production) and pretty printing (development).
*/
var CloudflareLogger = class CloudflareLogger {
	baseContext;
	minLevel;
	outputMode;
	/**
	* Create a new CloudflareLogger
	*
	* @param baseContext Base context included in all log entries
	* @param minLevel Minimum log level to output (default: INFO)
	* @param outputMode How log entries are formatted and emitted (default: 'structured')
	*/
	constructor(baseContext, minLevel = LogLevel.INFO, outputMode = "structured") {
		this.baseContext = baseContext;
		this.minLevel = minLevel;
		this.outputMode = outputMode;
	}
	/**
	* Log debug-level message
	*/
	debug(message, context) {
		if (this.shouldLog(LogLevel.DEBUG)) {
			const logData = this.buildLogData("debug", message, context);
			this.output(console.log, logData);
		}
	}
	/**
	* Log info-level message
	*/
	info(message, context) {
		if (this.shouldLog(LogLevel.INFO)) {
			const logData = this.buildLogData("info", message, context);
			this.output(console.log, logData);
		}
	}
	/**
	* Log warning-level message
	*/
	warn(message, context) {
		if (this.shouldLog(LogLevel.WARN)) {
			const logData = this.buildLogData("warn", message, context);
			this.output(console.warn, logData);
		}
	}
	/**
	* Log error-level message
	*/
	error(message, error, context) {
		if (this.shouldLog(LogLevel.ERROR)) {
			const logData = this.buildLogData("error", message, context, error);
			this.output(console.error, logData);
		}
	}
	/**
	* Create a child logger with additional context
	*/
	child(context) {
		return new CloudflareLogger({
			...this.baseContext,
			...context
		}, this.minLevel, this.outputMode);
	}
	/**
	* Check if a log level should be output
	*/
	shouldLog(level) {
		return level >= this.minLevel;
	}
	/**
	* Build log data object
	*/
	buildLogData(level, message, context, error) {
		const logData = {
			level,
			message,
			...this.baseContext,
			...context,
			timestamp: (/* @__PURE__ */ new Date()).toISOString()
		};
		if (error) logData.error = {
			message: error.message,
			stack: error.stack,
			name: error.name
		};
		return logData;
	}
	/**
	* Output log data using the configured output mode
	*/
	output(consoleFn, data) {
		switch (this.outputMode) {
			case "pretty":
				this.outputPretty(consoleFn, data);
				break;
			case "json-line":
				this.outputJsonLine(consoleFn, data);
				break;
			case "structured":
				this.outputStructured(consoleFn, data);
				break;
		}
	}
	/**
	* Output as JSON string (container stdout — parsed by Containers pipeline)
	*/
	outputJsonLine(consoleFn, data) {
		consoleFn(JSON.stringify(data));
	}
	/**
	* Output as raw object (Workers/DOs — Workers Logs auto-indexes fields)
	*/
	outputStructured(consoleFn, data) {
		consoleFn(data);
	}
	/**
	* Output as pretty-printed, colored text (development)
	*
	* Each log event is a single consoleFn() call so it appears as one entry
	* in the Cloudflare dashboard. Context is rendered inline as compact key=value pairs.
	*
	* Format: LEVEL [component] message trace=tr_... key=value key=value
	*/
	outputPretty(consoleFn, data) {
		const { level, message: msg, timestamp, traceId, component, sandboxId, sessionId, processId, commandId, durationMs, serviceVersion, instanceId, error, ...rest } = data;
		const levelStr = String(level || "INFO").toUpperCase();
		const levelColor = this.getLevelColor(levelStr);
		const componentBadge = component ? `[${component}]` : "";
		let logLine = `${timestamp ? `${COLORS.dim}${new Date(timestamp).toISOString().substring(11, 23)}${COLORS.reset} ` : ""}${levelColor}${levelStr.padEnd(5)}${COLORS.reset} ${componentBadge} ${msg}`;
		const pairs = [];
		if (traceId) pairs.push(`trace=${String(traceId).substring(0, 12)}`);
		if (commandId) pairs.push(`cmd=${String(commandId).substring(0, 12)}`);
		if (sandboxId) pairs.push(`sandbox=${sandboxId}`);
		if (sessionId) pairs.push(`session=${String(sessionId).substring(0, 12)}`);
		if (processId) pairs.push(`proc=${processId}`);
		if (durationMs !== void 0) pairs.push(`dur=${durationMs}ms`);
		for (const [key, value] of Object.entries(rest)) {
			if (value === void 0 || value === null) continue;
			const v = typeof value === "object" ? JSON.stringify(value) : this.sanitizePrettyValue(String(value));
			pairs.push(`${key}=${v}`);
		}
		if (error && typeof error === "object") {
			const errorObj = error;
			if (errorObj.name) pairs.push(`err.name=${this.sanitizePrettyValue(errorObj.name)}`);
			if (errorObj.message) pairs.push(`err.msg=${this.sanitizePrettyValue(errorObj.message)}`);
			if (errorObj.stack) pairs.push(`err.stack=${this.sanitizePrettyValue(errorObj.stack)}`);
		}
		if (pairs.length > 0) logLine += ` ${COLORS.dim}${pairs.join(" ")}${COLORS.reset}`;
		consoleFn(logLine);
	}
	/**
	* Collapse newlines so a single consoleFn() call stays on one line.
	* Cloudflare's log pipeline splits on literal newlines, which fragments
	* stack traces and multi-line error messages into separate entries.
	*/
	sanitizePrettyValue(value) {
		return value.replace(/\r/g, "\\r").replace(/\n/g, "\\n");
	}
	/**
	* Get ANSI color code for log level
	*/
	getLevelColor(level) {
		switch (level.toLowerCase()) {
			case "debug": return COLORS.debug;
			case "info": return COLORS.info;
			case "warn": return COLORS.warn;
			case "error": return COLORS.error;
			default: return COLORS.reset;
		}
	}
};

//#endregion
//#region ../shared/dist/logger/trace-context.js
/**
* Trace context utilities for request correlation
*
* Trace IDs enable correlating logs across distributed components:
* Worker → Durable Object → Container → back
*
* The trace ID is propagated via the X-Trace-Id HTTP header.
*/
/**
* Utility for managing trace context across distributed components
*/
var TraceContext = class TraceContext {
	/**
	* HTTP header name for trace ID propagation
	*/
	static TRACE_HEADER = "X-Trace-Id";
	/**
	* Generate a new trace ID
	*
	* Format: "tr_" + 16 random hex characters
	* Example: "tr_7f3a9b2c4e5d6f1a"
	*
	* @returns Newly generated trace ID
	*/
	static generate() {
		return `tr_${crypto.randomUUID().replace(/-/g, "").substring(0, 16)}`;
	}
	/**
	* Extract trace ID from HTTP request headers
	*
	* @param headers Request headers
	* @returns Trace ID if present, null otherwise
	*/
	static fromHeaders(headers) {
		return headers.get(TraceContext.TRACE_HEADER);
	}
	/**
	* Create headers object with trace ID for outgoing requests
	*
	* @param traceId Trace ID to include
	* @returns Headers object with X-Trace-Id set
	*/
	static toHeaders(traceId) {
		return { [TraceContext.TRACE_HEADER]: traceId };
	}
	/**
	* Get the header name used for trace ID propagation
	*
	* @returns Header name ("X-Trace-Id")
	*/
	static getHeaderName() {
		return TraceContext.TRACE_HEADER;
	}
};

//#endregion
//#region ../shared/dist/logger/index.js
/**
* Logger module
*
* Provides structured, trace-aware logging with:
* - Explicit logger passing via constructor injection
* - Three output modes: structured (Workers/DOs), json-line (container), pretty (local dev)
* - Environment auto-detection
* - Log level configuration
*
* Usage:
*
* ```typescript
* // Create a logger at entry point
* const logger = createLogger({ component: 'sandbox-do', traceId: 'tr_abc123' });
*
* // Pass to classes via constructor
* const service = new MyService(logger);
*
* // Create child loggers for additional context
* const execLogger = logger.child({ commandId: 'cmd-456' });
* execLogger.info('Operation started');
* ```
*/
/**
* Create a no-op logger for testing
*
* Returns a logger that implements the Logger interface but does nothing.
* Useful for tests that don't need actual logging output.
*
* @returns No-op logger instance
*
* @example
* ```typescript
* // In tests
* const client = new HttpClient({
*   baseUrl: 'http://test.com',
*   logger: createNoOpLogger() // Optional - tests can enable real logging if needed
* });
* ```
*/
function createNoOpLogger() {
	return {
		debug: () => {},
		info: () => {},
		warn: () => {},
		error: () => {},
		child: () => createNoOpLogger()
	};
}
/**
* Create a new logger instance
*
* @param context Base context for the logger. Must include 'component'.
*                TraceId will be auto-generated if not provided.
* @returns New logger instance
*
* @example
* ```typescript
* // In Durable Object
* const logger = createLogger({
*   component: 'sandbox-do',
*   traceId: TraceContext.fromHeaders(request.headers) || TraceContext.generate(),
*   sandboxId: this.id
* });
*
* // In Container
* const logger = createLogger({
*   component: 'container',
*   traceId: TraceContext.fromHeaders(request.headers)!,
*   sessionId: this.id
* });
* ```
*/
function createLogger(context) {
	const minLevel = getLogLevelFromEnv();
	const outputMode = getOutputMode(context.component);
	return new CloudflareLogger({
		...context,
		traceId: context.traceId || TraceContext.generate(),
		component: context.component,
		serviceVersion: context.serviceVersion || getEnvVar("SANDBOX_VERSION") || void 0,
		instanceId: context.instanceId || getEnvVar("HOSTNAME") || getEnvVar("SANDBOX_INSTANCE_ID") || void 0
	}, minLevel, outputMode);
}
/**
* Get log level from environment variable
*
* Checks SANDBOX_LOG_LEVEL env var, falls back to default based on environment.
* Default: 'debug' for development, 'info' for production
*/
function getLogLevelFromEnv() {
	switch ((getEnvVar("SANDBOX_LOG_LEVEL") || "info").toLowerCase()) {
		case "debug": return LogLevel.DEBUG;
		case "info": return LogLevel.INFO;
		case "warn": return LogLevel.WARN;
		case "error": return LogLevel.ERROR;
		default: return LogLevel.INFO;
	}
}
/**
* Determine output mode based on component and environment:
* - SANDBOX_LOG_FORMAT=pretty → 'pretty' for all components (local wrangler dev)
* - Container/Executor without pretty → 'json-line' (Bun stdout → Containers pipeline)
* - Everything else without pretty → 'structured' (Workers/DOs → Workers Logs)
*
* In local dev, setting SANDBOX_LOG_FORMAT=pretty gives readable terminal
* output on both the DO side and container side. In production (where the
* var isn't set), DOs emit structured objects and containers emit single-line
* JSON — both queryable by their respective observability pipelines.
*/
function getOutputMode(component) {
	if (getEnvVar("SANDBOX_LOG_FORMAT")?.toLowerCase() === "pretty") return "pretty";
	if (component === "container" || component === "executor") return "json-line";
	return "structured";
}
/**
* Get environment variable value
*
* Supports both Node.js (process.env) and Bun (Bun.env)
*/
function getEnvVar(name) {
	if (typeof process !== "undefined" && process.env) return process.env[name];
	if (typeof Bun !== "undefined") {
		const bunEnv = Bun.env;
		if (bunEnv) return bunEnv[name];
	}
}

//#endregion
//#region ../shared/dist/shell-escape.js
/**
* Escapes a string for safe use in shell commands using POSIX single-quote escaping.
* Prevents command injection by wrapping the string in single quotes and escaping
* any single quotes within the string.
*/
function shellEscape(str) {
	return `'${str.replace(/'/g, "'\\''")}'`;
}

//#endregion
//#region ../shared/dist/sse.js
/**
* Shared SSE parsing utilities.
*
* Parses SSE frames from arbitrary text chunks while preserving partial state
* across chunk boundaries.
*/
/**
* Parse SSE frames from a buffer.
*
* Returns parsed events, remaining unparsed text, and the current partial event
* so callers can continue parsing on the next chunk.
*/
function parseSSEFrames(buffer, currentEvent = { data: [] }) {
	const events = [];
	let i = 0;
	while (i < buffer.length) {
		const newlineIndex = buffer.indexOf("\n", i);
		if (newlineIndex === -1) break;
		const rawLine = buffer.substring(i, newlineIndex);
		const line = rawLine.endsWith("\r") ? rawLine.slice(0, -1) : rawLine;
		i = newlineIndex + 1;
		if (line === "" && currentEvent.data.length > 0) {
			events.push({
				event: currentEvent.event,
				data: currentEvent.data.join("\n")
			});
			currentEvent = { data: [] };
			continue;
		}
		if (line.startsWith("event:")) {
			currentEvent.event = line.startsWith("event: ") ? line.substring(7) : line.substring(6);
			continue;
		}
		if (line.startsWith("data:")) {
			const value = line.startsWith("data: ") ? line.substring(6) : line.substring(5);
			currentEvent.data.push(value);
		}
	}
	return {
		events,
		remaining: buffer.substring(i),
		currentEvent
	};
}

//#endregion
//#region ../shared/dist/types.js
function isExecResult(value) {
	return value && typeof value.success === "boolean" && typeof value.exitCode === "number" && typeof value.stdout === "string" && typeof value.stderr === "string";
}
function isProcess(value) {
	return value && typeof value.id === "string" && typeof value.command === "string" && typeof value.status === "string";
}
function isProcessStatus(value) {
	return [
		"starting",
		"running",
		"completed",
		"failed",
		"killed",
		"error"
	].includes(value);
}

//#endregion
//#region ../shared/dist/ws-types.js
/**
* Type guard for WSResponse
*
* Note: Only validates the discriminator field (type === 'response').
*/
function isWSResponse(msg) {
	return typeof msg === "object" && msg !== null && "type" in msg && msg.type === "response";
}
/**
* Type guard for WSStreamChunk
*
* Note: Only validates the discriminator field (type === 'stream').
*/
function isWSStreamChunk(msg) {
	return typeof msg === "object" && msg !== null && "type" in msg && msg.type === "stream";
}
/**
* Type guard for WSError
*
* Note: Only validates the discriminator field (type === 'error').
*/
function isWSError(msg) {
	return typeof msg === "object" && msg !== null && "type" in msg && msg.type === "error";
}
/**
* Generate a unique request ID
*/
function generateRequestId() {
	return `ws_${Date.now()}_${Math.random().toString(36).substring(2, 10)}`;
}

//#endregion
export { extractRepoName as _, isExecResult as a, partitionEnvVars as b, parseSSEFrames as c, createNoOpLogger as d, TraceContext as f, GitLogger as g, ResultImpl as h, isWSStreamChunk as i, shellEscape as l, Execution as m, isWSError as n, isProcess as o, logCanonicalEvent as p, isWSResponse as r, isProcessStatus as s, generateRequestId as t, createLogger as u, filterEnvVars as v, getEnvString as y };
