// react-router-auto-routes/dist/cli.cjs
#!/usr/bin/env node
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// src/migration/cli/run-cli.ts
var import_node_child_process2 = require("child_process");
var import_node_fs7 = __toESM(require("fs"), 1);
var import_node_path8 = __toESM(require("path"), 1);

// src/migration/fs-helpers.ts
var import_node_fs2 = __toESM(require("fs"), 1);
var import_node_path2 = __toESM(require("path"), 1);

// src/fs/visit-files.ts
var import_node_fs = __toESM(require("fs"), 1);
var import_node_path = __toESM(require("path"), 1);
function visitFiles(dir, visitor, options = {}) {
  const { followSymlinks = true } = options;
  const rootDir = dir;
  const readStat = followSymlinks ? import_node_fs.default.statSync : import_node_fs.default.lstatSync;
  const walk = (currentDir) => {
    for (const filename of import_node_fs.default.readdirSync(currentDir)) {
      const file = import_node_path.default.resolve(currentDir, filename);
      const stat = readStat(file);
      if (stat.isDirectory()) {
        walk(file);
      } else if (stat.isFile()) {
        visitor(import_node_path.default.relative(rootDir, file));
      }
    }
  };
  walk(dir);
}

// src/migration/logger.ts
function logInfo(message) {
  console.log(message);
}
function logWarn(message) {
  console.warn(message);
}
function logError(error) {
  if (error instanceof Error) {
    console.error(`\u274C ${error.message}`);
    return;
  }
  console.error(`\u274C ${String(error)}`);
}

// src/migration/fs-helpers.ts
function visitFiles2(dir, visitor) {
  visitFiles(dir, visitor, { followSymlinks: false });
}
var ROUTE_ENTRY_BASENAMES = /* @__PURE__ */ new Set(["route", "index", "_index", "_layout"]);
function isRouteEntryBasename(basename) {
  const ext = import_node_path2.default.extname(basename);
  const name = ext ? basename.slice(0, -ext.length) : basename;
  if (name.startsWith("__")) {
    return false;
  }
  return ROUTE_ENTRY_BASENAMES.has(name) || name.startsWith("_");
}
function isRouteSegment(segment) {
  if (segment === "" || segment === ".") return true;
  if (segment.endsWith("+")) return true;
  if (segment.startsWith("(") && segment.endsWith(")")) return true;
  if (segment.startsWith("__")) return true;
  if (segment.startsWith("_")) return true;
  return false;
}
function isColocatedFile(filename) {
  const normalized = filename.replace(/\\/g, "/");
  const segments = normalized.split("/");
  if (segments.length <= 1) return false;
  const directorySegments = segments.slice(0, -1);
  const basename = segments[segments.length - 1];
  const usesFlatFilesConvention = directorySegments.some(
    (segment) => segment.endsWith("+")
  );
  if (!usesFlatFilesConvention) return false;
  if (basename.startsWith("__")) {
    return true;
  }
  let lastPlusIndex = -1;
  for (let i = directorySegments.length - 1; i >= 0; i--) {
    if (directorySegments[i].endsWith("+")) {
      lastPlusIndex = i;
      break;
    }
  }
  const segmentsAfterPlus = directorySegments.slice(lastPlusIndex + 1);
  const regularSegments = segmentsAfterPlus.filter(
    (segment) => !isRouteSegment(segment)
  );
  if (regularSegments.length <= 1 && isRouteEntryBasename(basename)) {
    return false;
  }
  return regularSegments.length > 0;
}
function defaultTargetDir(sourceDir) {
  const absolute = import_node_path2.default.resolve(sourceDir);
  const parent = import_node_path2.default.dirname(absolute);
  const fallback = import_node_path2.default.join(parent, "new-routes");
  return pathRelative(process.cwd(), fallback);
}
function swapRoutes(sourceDir, targetDir, backupDir) {
  if (!import_node_fs2.default.existsSync(targetDir)) {
    throw new Error(`target directory '${targetDir}' was not created`);
  }
  const workingDir = process.cwd();
  const sourceRelative = pathRelative(workingDir, sourceDir);
  const targetRelative = pathRelative(workingDir, targetDir);
  const backupRelative = pathRelative(workingDir, backupDir);
  import_node_fs2.default.renameSync(sourceDir, backupDir);
  logInfo(`\u{1F4E6} Backed up '${sourceRelative}' to '${backupRelative}'.`);
  import_node_fs2.default.renameSync(targetDir, sourceDir);
  logInfo(`\u{1F680} Promoted '${targetRelative}' to '${sourceRelative}'.`);
}
function revertRoutes(sourceDir, targetDir, backupDir) {
  const workingDir = process.cwd();
  const sourceRelative = pathRelative(workingDir, sourceDir);
  const targetRelative = pathRelative(workingDir, targetDir);
  const backupRelative = pathRelative(workingDir, backupDir);
  if (import_node_fs2.default.existsSync(sourceDir)) {
    import_node_fs2.default.renameSync(sourceDir, targetDir);
    logInfo(
      `\u21A9\uFE0F Returned '${sourceRelative}' back to '${targetRelative}' for review.`
    );
  }
  if (import_node_fs2.default.existsSync(backupDir)) {
    import_node_fs2.default.renameSync(backupDir, sourceDir);
    logInfo(
      `\u{1F504} Restored original routes from '${backupRelative}' to '${sourceRelative}'.`
    );
  }
}
function pathRelative(from, to) {
  const relative3 = import_node_path2.default.relative(from, to);
  return relative3 === "" ? "." : relative3;
}
function normalizeDirectoryPath(dir) {
  if (dir === "") {
    return dir;
  }
  let normalized = import_node_path2.default.normalize(dir);
  if (normalized !== import_node_path2.default.sep) {
    normalized = normalized.replace(/[\\/]+$/, "");
  }
  if (normalized === ".") {
    return normalized;
  }
  const dotSlashPrefix = `.${import_node_path2.default.sep}`;
  if (normalized.startsWith(dotSlashPrefix)) {
    normalized = normalized.slice(dotSlashPrefix.length);
  }
  return normalized;
}

// src/migration/migrate.ts
var path8 = __toESM(require("path"), 1);

// src/migration/constants.ts
var MIGRATION_ROUTE_EXTENSIONS = Object.freeze([
  ".js",
  ".jsx",
  ".ts",
  ".tsx",
  ".md",
  ".mdx"
]);
var DOT_INDEX_SUFFIX = ".index";
var DOT_UNDERSCORE_INDEX_SUFFIX = "._index";

// src/core/routing/segments.ts
var import_path = __toESM(require("path"), 1);

// src/core/routing/constants.ts
var SPECIAL_ROUTE_FILES = [
  "index",
  "_index",
  "route",
  "_route",
  "_layout"
];

// src/core/routing/segments.ts
var ESCAPE_START = "[";
var ESCAPE_END = "]";
var OPTIONAL_START = "(";
var OPTIONAL_END = ")";
var WINDOWS_SEPARATOR = import_path.default.win32.sep;
var STRIP_ALWAYS = /* @__PURE__ */ new Set(["route", "_route"]);
function isSegmentSeparator(char) {
  if (!char) return false;
  return char === "/" || char === "." || char === WINDOWS_SEPARATOR;
}
function pushRouteSegment(routeSegments, rawRouteSegments, segment, rawSegment, routeId) {
  if (!segment) return;
  const checks = [
    [rawSegment, "*", rawSegment],
    [rawSegment, ":", rawSegment],
    [rawSegment, "/", segment]
  ];
  for (const [testValue, char, value] of checks) {
    if (testValue.includes(char)) {
      throw new Error(
        `Route segment "${value}" for "${routeId}" cannot contain "${char}".
If this is something you need, upvote this proposal for React Router https://github.com/remix-run/react-router/discussions/9822.`
      );
    }
  }
  routeSegments.push(segment);
  rawRouteSegments.push(rawSegment);
}
function stripTrailingSpecialSegments(originalName, routeSegments, rawRouteSegments) {
  if (!originalName.includes("/")) {
    return;
  }
  while (routeSegments.length > 0 && rawRouteSegments.length > 0) {
    const lastRaw = rawRouteSegments[rawRouteSegments.length - 1];
    if (!lastRaw || !STRIP_ALWAYS.has(lastRaw)) {
      break;
    }
    routeSegments.pop();
    rawRouteSegments.pop();
  }
}
function getRouteSegments(routeId, isIndex, paramChar = "$") {
  const routeSegments = [];
  const rawRouteSegments = [];
  let routeSegment = "";
  let rawRouteSegment = "";
  let state = "NORMAL";
  const flushSegment = () => {
    pushRouteSegment(
      routeSegments,
      rawRouteSegments,
      routeSegment,
      rawRouteSegment,
      routeId
    );
    routeSegment = "";
    rawRouteSegment = "";
  };
  for (let index = 0; index < routeId.length; index++) {
    const char = routeId[index];
    if (state === "ESCAPE" || state === "OPTIONAL_ESCAPE") {
      if (char === ESCAPE_END) {
        state = state === "ESCAPE" ? "NORMAL" : "OPTIONAL";
        rawRouteSegment += char;
      } else {
        routeSegment += char;
        rawRouteSegment += char;
      }
      continue;
    }
    if (state === "NORMAL") {
      if (isSegmentSeparator(char)) {
        flushSegment();
        continue;
      }
      if (char === ESCAPE_START) {
        state = "ESCAPE";
        rawRouteSegment += char;
        continue;
      }
      if (char === OPTIONAL_START) {
        state = "OPTIONAL";
        rawRouteSegment += char;
        continue;
      }
    }
    if (state === "OPTIONAL") {
      if (char === OPTIONAL_END) {
        routeSegment += "?";
        rawRouteSegment += char;
        state = "NORMAL";
        continue;
      }
      if (char === ESCAPE_START) {
        state = "OPTIONAL_ESCAPE";
        rawRouteSegment += char;
        continue;
      }
    }
    if (!routeSegment && char === paramChar && (state === "NORMAL" || state === "OPTIONAL")) {
      const next = routeId[index + 1];
      routeSegment += index + 1 === routeId.length || isSegmentSeparator(next) ? "*" : ":";
      rawRouteSegment += char;
      continue;
    }
    routeSegment += char;
    rawRouteSegment += char;
  }
  flushSegment();
  stripTrailingSpecialSegments(routeId, routeSegments, rawRouteSegments);
  const path12 = createRoutePath(routeSegments, rawRouteSegments, isIndex);
  const segments = deriveNameSegments(routeId, rawRouteSegments);
  return { segments, path: path12 };
}
function createRoutePath(routeSegments, rawRouteSegments, isIndex) {
  const length = isIndex ? routeSegments.length - 1 : routeSegments.length;
  if (length <= 0) {
    return void 0;
  }
  const parts = [];
  for (let i = 0; i < length; i++) {
    let segment = routeSegments[i];
    const rawSegment = rawRouteSegments[i];
    if (!rawSegment) {
      continue;
    }
    if (segment.startsWith("_") && rawSegment.startsWith("_")) {
      continue;
    }
    if (segment.endsWith("_") && rawSegment.endsWith("_")) {
      segment = segment.slice(0, -1);
    }
    parts.push(segment);
  }
  return parts.length > 0 ? parts.join("/") : void 0;
}
function deriveNameSegments(routeId, rawSegments) {
  const segments = [...rawSegments];
  while (segments.length > 0) {
    const last = segments[segments.length - 1];
    if (!last || !SPECIAL_ROUTE_FILES.includes(last)) {
      break;
    }
    segments.pop();
  }
  return segments;
}

// src/migration/route-scanner.ts
var import_node_fs3 = __toESM(require("fs"), 1);
var import_node_path3 = __toESM(require("path"), 1);
var import_picomatch = __toESM(require("picomatch"), 1);

// src/utils.ts
var path4 = __toESM(require("path"), 1);
function defaultVisitFiles(dir, visitor) {
  visitFiles(dir, visitor);
}
function createRouteId(file) {
  const normalized = file.split(path4.win32.sep).join("/");
  return normalized.replace(/\.[a-z0-9]+$/i, "");
}

// src/migration/route-scanner.ts
var ignoredFilenames = /* @__PURE__ */ new Set([".DS_Store"]);
function scanRouteModules(options) {
  const { appDirectory, routesDirectory, ignoredFilePatterns } = options;
  const appRoutesDirectory = import_node_path3.default.join(appDirectory, routesDirectory);
  if (!import_node_fs3.default.existsSync(appRoutesDirectory)) {
    throw new Error(`Routes directory not found: ${appRoutesDirectory}`);
  }
  const files = {};
  defaultVisitFiles(appRoutesDirectory, (file) => {
    const basename = import_node_path3.default.basename(file);
    if (ignoredFilenames.has(basename)) {
      return;
    }
    if (ignoredFilePatterns.length > 0 && ignoredFilePatterns.some(
      (pattern) => import_picomatch.default.isMatch(file, pattern, { dot: true })
    )) {
      return;
    }
    if (isColocatedFile(file)) {
      return;
    }
    const extension = import_node_path3.default.extname(file);
    if (MIGRATION_ROUTE_EXTENSIONS.includes(extension)) {
      const relativePath = import_node_path3.default.join(routesDirectory, file);
      const basename2 = import_node_path3.default.basename(file, extension);
      let routeId;
      if (basename2 === "route") {
        routeId = import_node_path3.default.dirname(relativePath).split(import_node_path3.default.win32.sep).join("/");
      } else {
        routeId = createRouteId(relativePath);
      }
      files[routeId] = relativePath;
      return;
    }
    throw new Error(
      `Invalid route module file: ${import_node_path3.default.join(appRoutesDirectory, file)}`
    );
  });
  return files;
}

// src/migration/create-routes-from-folders.ts
function createRoutesFromFolders(defineRoutes2, options = {}) {
  let {
    appDirectory = "app",
    ignoredFilePatterns = [],
    routesDirectory = "routes"
  } = options;
  const files = scanRouteModules({
    appDirectory,
    routesDirectory,
    ignoredFilePatterns
  });
  let routeIds = Object.keys(files).sort(byLongestFirst);
  let parentRouteIds = getParentRouteIds(routeIds);
  let seenRoutes = /* @__PURE__ */ new Map();
  function defineNestedRoutes(defineRoute, parentId) {
    let childRouteIds = routeIds.filter((id) => {
      return parentRouteIds[id] === parentId;
    });
    for (let routeId of childRouteIds) {
      let routePath = createRoutePath2(
        routeId.slice((parentId || routesDirectory).length + 1)
      );
      let isIndexRoute = isIndexRouteId(routeId);
      let fullPath = createRoutePath2(routeId.slice(routesDirectory.length + 1));
      let uniqueRouteId = (fullPath || "") + (isIndexRoute ? "?index" : "");
      let isPathlessLayoutRoute = routeId.split("/").pop()?.startsWith("__") === true;
      ensureUniqueRoute(seenRoutes, {
        id: routeId,
        fullPath: fullPath || "/",
        uniqueKey: uniqueRouteId,
        isPathlessLayoutRoute
      });
      if (isIndexRoute) {
        let invalidChildRoutes = routeIds.filter(
          (id) => parentRouteIds[id] === routeId
        );
        if (invalidChildRoutes.length > 0) {
          throw new Error(
            `Child routes are not allowed in index routes. Please remove child routes of ${routeId}`
          );
        }
        defineRoute(routePath, files[routeId], { index: true, id: routeId });
      } else {
        defineRoute(routePath, files[routeId], { id: routeId }, () => {
          defineNestedRoutes(defineRoute, routeId);
        });
      }
    }
  }
  return defineRoutes2(defineNestedRoutes);
}
function getParentRouteIds(routeIds) {
  const routeIdMap = /* @__PURE__ */ new Map();
  for (const routeId of routeIds) {
    routeIdMap.set(routeId, routeId);
  }
  const parentRouteIdMap = /* @__PURE__ */ new Map();
  for (const [childRouteId, _] of routeIdMap) {
    let parentRouteId = void 0;
    for (const [potentialParentId, _2] of routeIdMap) {
      if (childRouteId.startsWith(`${potentialParentId}/`)) {
        parentRouteId = potentialParentId;
        break;
      }
    }
    if (!parentRouteId && childRouteId.endsWith(DOT_INDEX_SUFFIX)) {
      const dotNotationParentId = childRouteId.slice(
        0,
        -DOT_INDEX_SUFFIX.length
      );
      if (routeIdMap.has(dotNotationParentId)) {
        parentRouteId = dotNotationParentId;
      }
    }
    if (!parentRouteId && childRouteId.endsWith(DOT_UNDERSCORE_INDEX_SUFFIX)) {
      const dotNotationParentId = childRouteId.slice(
        0,
        -DOT_UNDERSCORE_INDEX_SUFFIX.length
      );
      if (routeIdMap.has(dotNotationParentId)) {
        parentRouteId = dotNotationParentId;
      }
    }
    parentRouteIdMap.set(childRouteId, parentRouteId);
  }
  return Object.fromEntries(parentRouteIdMap);
}
function byLongestFirst(a, b) {
  return b.length - a.length;
}
function ensureUniqueRoute(seen, context) {
  if (!context.uniqueKey || context.isPathlessLayoutRoute) {
    return;
  }
  const conflict = seen.get(context.uniqueKey);
  if (conflict) {
    throw new Error(
      `Path ${JSON.stringify(context.fullPath)} defined by route ${JSON.stringify(context.id)} conflicts with route ${JSON.stringify(conflict)}`
    );
  }
  seen.set(context.uniqueKey, context.id);
}
function createRoutePath2(partialRouteId) {
  validatePartialRoute(partialRouteId);
  if (partialRouteId.endsWith("/_layout")) {
    const parentId = partialRouteId.slice(0, -"/_layout".length);
    const parentPath = createRoutePath2(parentId);
    if (!parentPath) {
      return "_layout";
    }
    return `${parentPath}/_layout`;
  }
  const { path: path12 } = getRouteSegments(
    partialRouteId,
    isIndexRouteId(partialRouteId)
  );
  if (!path12) {
    return path12;
  }
  if (/^\(__[^)]+\)\//.test(partialRouteId) && path12.startsWith("__") && path12.includes("?/")) {
    const segments = path12.split("/");
    const [first, ...rest] = segments;
    if (first?.startsWith("__") && first.endsWith("?")) {
      const remainder = rest.join("/");
      return remainder ? `(${remainder}` : "(";
    }
  }
  return path12;
}
function isIndexRouteId(routeId) {
  return routeId === "index" || routeId.endsWith("/index") || routeId.endsWith(DOT_INDEX_SUFFIX) || routeId.endsWith(DOT_UNDERSCORE_INDEX_SUFFIX);
}
function validatePartialRoute(partialRouteId) {
  if (/\(\$\)(?:$|[/.])/.test(partialRouteId)) {
    throw new Error(
      `Invalid route path: ${partialRouteId}. Splat route $ is already optional`
    );
  }
  if (/(^|[/.])\(index\)(?:$|[/.])/.test(partialRouteId)) {
    throw new Error(
      `Invalid route path: ${partialRouteId}. Make index route optional by using (index)`
    );
  }
}

// src/migration/import-rewriter.ts
var import_node_fs4 = __toESM(require("fs"), 1);
var import_node_path4 = __toESM(require("path"), 1);
var import_typescript = __toESM(require("typescript"), 1);
var transformableExtensions = /* @__PURE__ */ new Set([".js", ".jsx", ".ts", ".tsx"]);
var resolutionExtensions = [
  ".ts",
  ".tsx",
  ".mts",
  ".cts",
  ".js",
  ".jsx",
  ".mjs",
  ".cjs",
  ".json",
  ".md",
  ".mdx"
];
function buildImportRewritePlan(mapping, normalizedMapping, specifierReplacements) {
  const sourcePath = normalizeAbsolutePath(mapping.source);
  const targetPath = normalizeAbsolutePath(mapping.target);
  const context = {
    sourcePath,
    targetPath,
    normalizedMapping,
    specifierReplacements
  };
  if (!shouldRewriteImports(sourcePath)) {
    return {
      mode: "copy",
      sourcePath,
      targetPath
    };
  }
  const originalContents = import_node_fs4.default.readFileSync(sourcePath, "utf8");
  const rewrittenContents = rewriteModuleImports(originalContents, context);
  return {
    mode: "rewrite",
    sourcePath,
    targetPath,
    originalContents,
    rewrittenContents,
    changed: originalContents !== rewrittenContents
  };
}
function executeImportRewritePlan(plan) {
  import_node_fs4.default.mkdirSync(import_node_path4.default.dirname(plan.targetPath), { recursive: true });
  if (plan.mode === "copy") {
    import_node_fs4.default.copyFileSync(plan.sourcePath, plan.targetPath);
    return;
  }
  import_node_fs4.default.writeFileSync(plan.targetPath, plan.rewrittenContents);
}
function rewriteAndCopy(mapping, normalizedMapping, specifierReplacements) {
  const plan = buildImportRewritePlan(
    mapping,
    normalizedMapping,
    specifierReplacements
  );
  executeImportRewritePlan(plan);
}
function normalizeAbsolutePath(filePath) {
  return import_node_path4.default.normalize(import_node_path4.default.resolve(filePath));
}
function shouldRewriteImports(filePath) {
  const ext = import_node_path4.default.extname(filePath).toLowerCase();
  return transformableExtensions.has(ext);
}
function rewriteModuleImports(contents, context) {
  const scriptKind = getScriptKindForFile(context.sourcePath);
  const sourceFile = import_typescript.default.createSourceFile(
    context.sourcePath,
    contents,
    import_typescript.default.ScriptTarget.Latest,
    true,
    scriptKind
  );
  const edits = [];
  const handleLiteral = (literal) => {
    const replacement = computeSpecifierReplacement(literal.text, context);
    if (!replacement) {
      return;
    }
    const literalText = literal.getText(sourceFile);
    const quote = literalText[0] === "`" || literalText[0] === '"' || literalText[0] === "'" ? literalText[0] : '"';
    edits.push({
      start: literal.getStart(sourceFile),
      end: literal.getEnd(),
      text: `${quote}${replacement}${quote}`
    });
  };
  const visit = (node) => {
    if (import_typescript.default.isImportDeclaration(node) && import_typescript.default.isStringLiteralLike(node.moduleSpecifier)) {
      handleLiteral(node.moduleSpecifier);
    } else if (import_typescript.default.isExportDeclaration(node) && node.moduleSpecifier && import_typescript.default.isStringLiteralLike(node.moduleSpecifier)) {
      handleLiteral(node.moduleSpecifier);
    } else if (import_typescript.default.isCallExpression(node) && node.arguments.length === 1 && import_typescript.default.isStringLiteralLike(node.arguments[0]) && (node.expression.kind === import_typescript.default.SyntaxKind.ImportKeyword || import_typescript.default.isIdentifier(node.expression) && node.expression.text === "require")) {
      handleLiteral(node.arguments[0]);
    }
    import_typescript.default.forEachChild(node, visit);
  };
  visit(sourceFile);
  if (edits.length === 0) {
    return contents;
  }
  edits.sort((a, b) => b.start - a.start);
  let result = contents;
  for (const edit of edits) {
    result = result.slice(0, edit.start) + edit.text + result.slice(edit.end);
  }
  return result;
}
function computeSpecifierReplacement(specifier, context) {
  const { base, suffix } = splitImportSpecifier(specifier);
  const relativeReplacement = rewriteRelativeSpecifier(base, context);
  if (relativeReplacement) {
    const stripped = stripTsExtension(relativeReplacement) ?? relativeReplacement;
    const nextSpecifier = stripped + suffix;
    if (nextSpecifier !== specifier) {
      return nextSpecifier;
    }
  }
  let nextBase = base;
  let changed = false;
  const typesReplacement = rewriteTypesRouteSpecifier(nextBase, context);
  if (typesReplacement) {
    nextBase = typesReplacement;
    changed = true;
  }
  const extensionReplacement = stripTsExtension(nextBase);
  if (extensionReplacement) {
    nextBase = extensionReplacement;
    changed = true;
  }
  const aliasedReplacement = rewriteAliasedSpecifier(nextBase, context);
  if (aliasedReplacement) {
    nextBase = aliasedReplacement;
    changed = true;
  }
  const legacyReplacement = rewriteLegacyPlusSegments(nextBase);
  if (legacyReplacement) {
    nextBase = legacyReplacement;
    changed = true;
  }
  if (changed) {
    const nextSpecifier = nextBase + suffix;
    if (nextSpecifier !== specifier) {
      return nextSpecifier;
    }
  }
  return null;
}
function rewriteTypesRouteSpecifier(specifier, context) {
  const normalized = specifier.replace(/\\/g, "/");
  const typesRoutePattern = /\/\+types\/route$/;
  if (!typesRoutePattern.test(normalized)) {
    return null;
  }
  const sourceBasename = import_node_path4.default.basename(context.sourcePath);
  const targetBasename = import_node_path4.default.basename(context.targetPath);
  const sourceNameWithoutExt = sourceBasename.replace(/\.[^.]+$/, "");
  const targetNameWithoutExt = targetBasename.replace(/\.[^.]+$/, "");
  if (sourceNameWithoutExt !== "route" || targetNameWithoutExt === "route") {
    return null;
  }
  return normalized.replace(
    typesRoutePattern,
    `/+types/${targetNameWithoutExt}`
  );
}
function stripTsExtension(specifier) {
  const normalized = specifier.replace(/\\/g, "/");
  const indexTsPattern = /\/index\.tsx?$/;
  if (indexTsPattern.test(normalized)) {
    return normalized.replace(indexTsPattern, "");
  }
  return null;
}
function splitImportSpecifier(specifier) {
  const queryIndex = specifier.indexOf("?");
  const hashIndex = specifier.startsWith("#") ? specifier.indexOf("#", 1) : specifier.indexOf("#");
  let cutIndex = specifier.length;
  const candidates = [queryIndex, hashIndex].filter((index) => index > -1);
  if (candidates.length > 0) {
    cutIndex = Math.min(...candidates);
  }
  return {
    base: specifier.slice(0, cutIndex),
    suffix: specifier.slice(cutIndex)
  };
}
function isRelativeSpecifier(specifier) {
  if (!specifier) {
    return false;
  }
  return specifier === "." || specifier === ".." || specifier.startsWith("./") || specifier.startsWith("../");
}
function rewriteRelativeSpecifier(specifier, context) {
  if (!isRelativeSpecifier(specifier)) {
    return null;
  }
  const resolution = resolveRelativeSpecifier(specifier, context.sourcePath);
  if (!resolution) {
    return null;
  }
  const resolvedAbsolute = normalizeAbsolutePath(resolution.resolvedPath);
  const migratedAbsolute = context.normalizedMapping.get(resolvedAbsolute) ?? resolvedAbsolute;
  const relativeSpecifier = computeRelativeSpecifier(
    migratedAbsolute,
    context.targetPath,
    specifier,
    resolution
  );
  return relativeSpecifier;
}
function rewriteLegacyPlusSegments(specifier) {
  if (!specifier.includes("+")) {
    return null;
  }
  const segments = specifier.split("/");
  let changed = false;
  let encounteredLegacyRoute = false;
  let colocatedIndex = -1;
  for (let index = 0; index < segments.length; index++) {
    const segment = segments[index];
    if (!segment || segment === "." || segment === "..") {
      continue;
    }
    if (segment.endsWith("+") && !segment.startsWith("+") && segment.length > 1 && segment[segment.length - 2] !== "+") {
      segments[index] = segment.slice(0, -1);
      changed = true;
      encounteredLegacyRoute = true;
      continue;
    }
    const isLastSegment = index === segments.length - 1;
    if (colocatedIndex === -1 && encounteredLegacyRoute && !isLastSegment && !segment.startsWith("+") && !segment.endsWith("+")) {
      colocatedIndex = index;
    }
  }
  if (colocatedIndex !== -1) {
    segments[colocatedIndex] = `+${segments[colocatedIndex]}`;
    changed = true;
  }
  if (!changed) {
    return null;
  }
  return segments.join("/");
}
function rewriteAliasedSpecifier(specifier, context) {
  if (context.specifierReplacements.length === 0) {
    return null;
  }
  const normalized = normalizeSpecifier(specifier);
  for (const replacement of context.specifierReplacements) {
    const index = normalized.lastIndexOf(replacement.from);
    if (index === -1) {
      continue;
    }
    const prefix = normalized.slice(0, index);
    const suffix = normalized.slice(index + replacement.from.length);
    const next = prefix + replacement.to + suffix;
    if (next !== normalized) {
      return next;
    }
  }
  return null;
}
function resolveRelativeSpecifier(specifier, fromFile) {
  const sourceDir = import_node_path4.default.dirname(fromFile);
  const basePath = import_node_path4.default.resolve(sourceDir, specifier);
  const baseStat = safeStat(basePath);
  if (baseStat?.isFile()) {
    return {
      resolvedPath: basePath,
      appendedExtension: "",
      usedIndex: false
    };
  }
  if (baseStat?.isDirectory()) {
    const indexFile = resolveIndexFile(basePath);
    if (indexFile) {
      return {
        resolvedPath: indexFile,
        appendedExtension: "",
        usedIndex: true
      };
    }
  }
  for (const extension of resolutionExtensions) {
    const candidate = basePath + extension;
    const stat = safeStat(candidate);
    if (stat?.isFile()) {
      return {
        resolvedPath: candidate,
        appendedExtension: extension,
        usedIndex: false
      };
    }
  }
  return null;
}
function resolveIndexFile(dir) {
  const noExtensionIndex = import_node_path4.default.join(dir, "index");
  const noExtStat = safeStat(noExtensionIndex);
  if (noExtStat?.isFile()) {
    return noExtensionIndex;
  }
  for (const extension of resolutionExtensions) {
    const candidate = import_node_path4.default.join(dir, `index${extension}`);
    const stat = safeStat(candidate);
    if (stat?.isFile()) {
      return candidate;
    }
  }
  return null;
}
function computeRelativeSpecifier(resolvedAbsolute, targetFile, originalBase, resolution) {
  const targetDir = import_node_path4.default.dirname(targetFile);
  let relativePath = import_node_path4.default.relative(targetDir, resolvedAbsolute);
  if (!relativePath) {
    return originalBase;
  }
  relativePath = relativePath.replace(/\\/g, "/");
  if (!relativePath.startsWith(".")) {
    relativePath = `./${relativePath}`;
  }
  if (resolution.appendedExtension && relativePath.endsWith(resolution.appendedExtension)) {
    relativePath = relativePath.slice(0, -resolution.appendedExtension.length);
  }
  const normalizedOriginal = originalBase.replace(/\\/g, "/");
  if ((resolution.usedIndex || !normalizedOriginal.endsWith("/index")) && relativePath.endsWith("/index")) {
    relativePath = relativePath.slice(0, -"/index".length);
  }
  if (!relativePath.startsWith(".")) {
    relativePath = `./${relativePath}`;
  }
  return relativePath;
}
function safeStat(filePath) {
  try {
    return import_node_fs4.default.statSync(filePath);
  } catch {
    return null;
  }
}
function normalizeSpecifier(value) {
  return value.replace(/\\/g, "/");
}
function getScriptKindForFile(filePath) {
  const ext = import_node_path4.default.extname(filePath).toLowerCase();
  switch (ext) {
    case ".tsx":
      return import_typescript.default.ScriptKind.TSX;
    case ".ts":
      return import_typescript.default.ScriptKind.TS;
    case ".jsx":
      return import_typescript.default.ScriptKind.JSX;
    case ".js":
      return import_typescript.default.ScriptKind.JS;
    default:
      return import_typescript.default.ScriptKind.TS;
  }
}

// src/migration/normalizers.ts
var import_node_path5 = __toESM(require("path"), 1);
function cleanLegacySegment(segment, dropTrailingPlus = true) {
  const withoutPlus = dropTrailingPlus ? segment.replace(/\+$/g, "") : segment;
  return withoutPlus.replace(
    /(^|\.|_)__+/g,
    (_match, prefix) => `${prefix}_`
  );
}
function convertToRoute(routes, sourceDir, id, parentId) {
  let routeId = id.substring(sourceDir.length + 1);
  parentId = parentId === "root" ? parentId : parentId.substring(sourceDir.length + 1);
  const segments = routeId.split("/");
  const normalizedSegments = segments.map(
    (segment, index) => cleanLegacySegment(segment, index < segments.length - 1)
  );
  let flat = normalizedSegments.join("/");
  if (Object.values(routes).some((route) => route.parentId === id)) {
    flat = flat + "/_layout";
  }
  return flat;
}
function convertColocatedPath(file) {
  const normalized = file.replace(/\\/g, "/");
  const segments = normalized.split("/");
  let colocatedFolderIndex = -1;
  for (let index = 0; index < segments.length - 1; index++) {
    const segment = segments[index];
    if (segment && !segment.endsWith("+") && segment !== "." && segment !== "") {
      colocatedFolderIndex = index;
      break;
    }
  }
  const converted = segments.map((segment, index) => {
    const isLastSegment = index === segments.length - 1;
    if (index === colocatedFolderIndex) {
      return "+" + segment;
    }
    if (!isLastSegment) {
      return segment.replace(/\+$/, "");
    }
    return segment;
  });
  return converted.join(import_node_path5.default.sep);
}
function normalizeSnapshotRouteFilePath(filePath) {
  const normalized = filePath.replace(/\\/g, "/");
  const ext = import_node_path5.default.extname(normalized);
  const base = ext ? normalized.slice(0, -ext.length) : normalized;
  const rawSegments = base.split("/");
  const result = [];
  for (const rawSegment of rawSegments) {
    const segment = cleanLegacySegment(rawSegment);
    if (!segment) {
      result.push(segment);
      continue;
    }
    if (segment === "_layout" && result.length > 0) {
      continue;
    }
    if (segment === "route" && result.length > 0) {
      continue;
    }
    if (segment === "index" && result.length > 0) {
      const parent = result.pop();
      result.push(`${parent}.index`);
      continue;
    }
    if (result.length > 0 && result[result.length - 1].endsWith("_") && segment.startsWith("_")) {
      const previous = result.pop();
      const suffix = segment.slice(1);
      const combined = suffix.length > 0 ? `${previous}.${suffix}` : previous;
      result.push(combined);
      continue;
    }
    result.push(segment);
  }
  return result.join("/") + ext;
}

// src/migration/route-definition.ts
function defineRoutes(callback) {
  let routes = /* @__PURE__ */ Object.create(null);
  let parentRoutes = [];
  let alreadyReturned = false;
  let defineRoute = (path12, file, optionsOrChildren, children) => {
    if (alreadyReturned) {
      throw new Error(
        "You tried to define routes asynchronously but started defining routes before the async work was done. Please await all async data before calling `defineRoutes()`"
      );
    }
    let options;
    if (typeof optionsOrChildren === "function") {
      options = {};
      children = optionsOrChildren;
    } else {
      options = optionsOrChildren || {};
    }
    let route = {
      path: path12 ? path12 : void 0,
      index: options.index ? true : void 0,
      caseSensitive: options.caseSensitive ? true : void 0,
      id: options.id || createRouteId(file),
      parentId: parentRoutes.length > 0 ? parentRoutes[parentRoutes.length - 1].id : "root",
      file
    };
    if (route.id in routes) {
      throw new Error(
        `Unable to define routes with duplicate route id: "${route.id}"`
      );
    }
    routes[route.id] = route;
    if (children) {
      parentRoutes.push(route);
      children();
      parentRoutes.pop();
    }
  };
  callback(defineRoute);
  alreadyReturned = true;
  return routes;
}

// src/migration/migrate.ts
function migrate(sourceDir, targetDir, options = {
  force: false,
  ignoredRouteFiles: void 0
}) {
  sourceDir = normalizeDirectoryPath(sourceDir);
  targetDir = normalizeDirectoryPath(targetDir);
  const ignoredPatterns = Array.from(
    /* @__PURE__ */ new Set([...options.ignoredRouteFiles ?? [], "**/.DS_Store"])
  );
  logInfo("\u{1F6E0}\uFE0F Migrating routes to + folder convention...");
  logInfo(`\u{1F5C2}\uFE0F source: ${sourceDir}`);
  logInfo(`\u{1F5C2}\uFE0F target: ${targetDir}`);
  logInfo(`\u{1F648}ignored files: ${ignoredPatterns}`);
  logInfo("");
  const routes = createRoutesFromFolders(defineRoutes, {
    appDirectory: "./",
    routesDirectory: sourceDir,
    ignoredFilePatterns: ignoredPatterns
  });
  const routeMappings = collectRouteMappings(routes, sourceDir, targetDir);
  const routeSourcePaths = new Set(
    routeMappings.map((m) => normalizeAbsolutePath(m.source))
  );
  const colocatedMappings = collectColocatedMappings(
    sourceDir,
    targetDir,
    routeSourcePaths
  );
  const mappings = [...routeMappings, ...colocatedMappings];
  const normalizedMapping = createNormalizedMapping(mappings);
  const specifierReplacements = createSpecifierReplacements(
    mappings,
    sourceDir,
    targetDir
  );
  for (const mapping of mappings) {
    rewriteAndCopy(mapping, normalizedMapping, specifierReplacements);
  }
  logInfo("\u{1F3C1} Finished!");
}
function collectRouteMappings(routes, sourceDir, targetDir) {
  const mappings = [];
  for (const [id, route] of Object.entries(routes)) {
    const { path: routePath, file, parentId } = route;
    const extension = path8.extname(file);
    if (!MIGRATION_ROUTE_EXTENSIONS.includes(extension)) {
      continue;
    }
    const flat = convertToRoute(routes, sourceDir, id, parentId);
    const sourcePath = path8.resolve(file);
    const targetPath = path8.resolve(targetDir, `${flat}${extension}`);
    mappings.push({ source: sourcePath, target: targetPath });
  }
  return mappings;
}
function collectColocatedMappings(sourceDir, targetDir, routeSourcePaths) {
  const mappings = [];
  visitFiles2(sourceDir, (file) => {
    if (!isColocatedFile(file)) {
      return;
    }
    const sourcePath = path8.resolve(sourceDir, file);
    if (routeSourcePaths.has(normalizeAbsolutePath(sourcePath))) {
      return;
    }
    const targetPath = path8.resolve(targetDir, convertColocatedPath(file));
    mappings.push({ source: sourcePath, target: targetPath });
  });
  return mappings;
}
function createNormalizedMapping(mappings) {
  const normalized = /* @__PURE__ */ new Map();
  for (const mapping of mappings) {
    normalized.set(
      normalizeAbsolutePath(mapping.source),
      normalizeAbsolutePath(mapping.target)
    );
  }
  return normalized;
}
function createSpecifierReplacements(mappings, sourceDir, targetDir) {
  if (mappings.length === 0) {
    return [];
  }
  const replacements = [];
  const seen = /* @__PURE__ */ new Set();
  const sourceDirAbsolute = normalizeAbsolutePath(path8.resolve(sourceDir));
  const targetDirAbsolute = normalizeAbsolutePath(path8.resolve(targetDir));
  const sourceDirImport = normalizeImportPath(sourceDir);
  const prefixes = createPrefixCandidates(sourceDirImport);
  for (const mapping of mappings) {
    const sourceAbsolute = normalizeAbsolutePath(mapping.source);
    const targetAbsolute = normalizeAbsolutePath(mapping.target);
    const sourceRelative = normalizeImportPath(
      path8.relative(sourceDirAbsolute, sourceAbsolute)
    );
    if (sourceRelative.startsWith("../")) {
      continue;
    }
    const targetRelative = normalizeImportPath(
      path8.relative(targetDirAbsolute, targetAbsolute)
    );
    if (targetRelative.startsWith("../")) {
      continue;
    }
    for (const prefix of prefixes) {
      const from = prefix ? `${prefix}/${sourceRelative}` : sourceRelative;
      const to = prefix ? `${prefix}/${targetRelative}` : targetRelative;
      if (from === to) {
        continue;
      }
      const key = `${from}->${to}`;
      if (seen.has(key)) {
        continue;
      }
      seen.add(key);
      replacements.push({ from, to });
    }
  }
  replacements.sort((a, b) => b.from.length - a.from.length);
  return replacements;
}
function normalizeImportPath(value) {
  let normalized = value.replace(/\\/g, "/");
  if (normalized.startsWith("./")) {
    normalized = normalized.slice(2);
  }
  normalized = normalized.replace(/\/+/g, "/");
  if (normalized.endsWith("/")) {
    normalized = normalized.slice(0, -1);
  }
  return normalized;
}
function createPrefixCandidates(base) {
  const prefixes = /* @__PURE__ */ new Set();
  if (base) {
    prefixes.add(base);
    const segments = base.split("/");
    for (let index = 1; index < segments.length; index += 1) {
      const suffix = segments.slice(index).join("/");
      if (suffix) {
        prefixes.add(suffix);
      }
    }
  }
  prefixes.add("");
  return Array.from(prefixes);
}

// src/migration/cli/diff.ts
function normalizeSnapshot(snapshot) {
  const normalizedRoutes = [];
  const lines = snapshot.replace(/\r\n/g, "\n").split("\n");
  for (const line of lines) {
    const routeMatch = line.match(/<Route\b([^>]*)>/);
    if (!routeMatch) {
      continue;
    }
    const attributes = routeMatch[1];
    const fileMatch = attributes.match(/file="([^"]+)"/);
    if (!fileMatch) {
      continue;
    }
    const normalizedFile = normalizeSnapshotRouteFilePath(fileMatch[1]);
    const isIndex = /(^|\s)index(\s|$)/.test(attributes);
    const caseSensitiveMatch = attributes.match(
      /caseSensitive="?(true|false)"?/
    );
    const caseSensitive = caseSensitiveMatch?.[1] === "true";
    const parts = [
      `file=${normalizedFile}`,
      `index=${isIndex ? "1" : "0"}`,
      `caseSensitive=${caseSensitive ? "1" : "0"}`
    ];
    normalizedRoutes.push(parts.join("|"));
  }
  normalizedRoutes.sort();
  return normalizedRoutes.join("\n");
}
function diffSnapshots(before, after) {
  const beforeLines = before.split("\n");
  const afterLines = after.split("\n");
  const m = beforeLines.length;
  const n = afterLines.length;
  const dp = Array.from(
    { length: m + 1 },
    () => Array.from({ length: n + 1 }, () => 0)
  );
  for (let i2 = 1; i2 <= m; i2 += 1) {
    for (let j2 = 1; j2 <= n; j2 += 1) {
      if (beforeLines[i2 - 1] === afterLines[j2 - 1]) {
        dp[i2][j2] = dp[i2 - 1][j2 - 1] + 1;
      } else {
        dp[i2][j2] = Math.max(dp[i2 - 1][j2], dp[i2][j2 - 1]);
      }
    }
  }
  const diff = [];
  let i = m;
  let j = n;
  while (i > 0 && j > 0) {
    if (beforeLines[i - 1] === afterLines[j - 1]) {
      diff.push({ type: "same", line: beforeLines[i - 1] });
      i -= 1;
      j -= 1;
    } else if (dp[i - 1][j] >= dp[i][j - 1]) {
      diff.push({ type: "remove", line: beforeLines[i - 1] });
      i -= 1;
    } else {
      diff.push({ type: "add", line: afterLines[j - 1] });
      j -= 1;
    }
  }
  while (i > 0) {
    diff.push({ type: "remove", line: beforeLines[i - 1] });
    i -= 1;
  }
  while (j > 0) {
    diff.push({ type: "add", line: afterLines[j - 1] });
    j -= 1;
  }
  diff.reverse();
  const lines = [
    "--- react-router routes (before)",
    "+++ react-router routes (after)"
  ];
  for (const entry of diff) {
    if (entry.type === "same") {
      lines.push(` ${entry.line}`);
    } else if (entry.type === "remove") {
      lines.push(`-${entry.line}`);
    } else {
      lines.push(`+${entry.line}`);
    }
  }
  return lines.join("\n");
}

// src/migration/cli/route-entry.ts
var import_node_fs5 = __toESM(require("fs"), 1);
var import_node_path6 = __toESM(require("path"), 1);
var import_typescript2 = __toESM(require("typescript"), 1);
var LEGACY_PATTERNS = [
  "remix-flat-routes",
  "createRoutesFromFolders",
  "flatRoutes",
  "@react-router/remix-routes-option-adapter"
];
var NEXT_PATTERNS = ["react-router-auto-routes"];
var ENTRY_CANDIDATES = ["routes.ts", "routes.tsx", "routes.js", "routes.jsx"];
function detectLegacyRouteEntry(sourceDir) {
  const appDir = import_node_path6.default.dirname(sourceDir);
  for (const candidate of ENTRY_CANDIDATES) {
    const candidatePath = import_node_path6.default.join(appDir, candidate);
    if (!import_node_fs5.default.existsSync(candidatePath) || !import_node_fs5.default.statSync(candidatePath).isFile()) {
      continue;
    }
    let contents;
    try {
      contents = import_node_fs5.default.readFileSync(candidatePath, "utf8");
    } catch {
      continue;
    }
    if (NEXT_PATTERNS.some((pattern) => contents.includes(pattern))) {
      return { entryPath: candidatePath, isLegacy: false };
    }
    if (LEGACY_PATTERNS.some((pattern) => contents.includes(pattern))) {
      return { entryPath: candidatePath, isLegacy: true };
    }
  }
  return { entryPath: null, isLegacy: false };
}
function rewriteLegacyRouteEntry(entryPath) {
  let contents;
  try {
    contents = import_node_fs5.default.readFileSync(entryPath, "utf8");
  } catch {
    return { updated: false };
  }
  if (contents.includes("react-router-auto-routes")) {
    return { updated: false };
  }
  const legacyOptions = extractLegacyRouteOptions(entryPath, contents);
  const printer = import_typescript2.default.createPrinter({ newLine: import_typescript2.default.NewLineKind.LineFeed });
  const optionsText = createAutoRoutesOptionsText(printer, legacyOptions);
  const nextContents = createAutoRoutesEntry(optionsText);
  import_node_fs5.default.writeFileSync(entryPath, nextContents);
  return {
    updated: true,
    previousContents: contents
  };
}
function createAutoRoutesEntry(optionsText) {
  const header = "import { autoRoutes } from 'react-router-auto-routes'\nimport { type RouteConfig } from '@react-router/dev/routes'\n\n";
  if (!optionsText) {
    return `${header}export default autoRoutes() satisfies RouteConfig
`;
  }
  return `${header}export default autoRoutes(${optionsText}) satisfies RouteConfig
`;
}
function extractLegacyRouteOptions(entryPath, contents) {
  const scriptKind = getScriptKind(entryPath);
  const sourceFile = import_typescript2.default.createSourceFile(
    entryPath,
    contents,
    import_typescript2.default.ScriptTarget.Latest,
    true,
    scriptKind
  );
  const legacyCalls = /* @__PURE__ */ new Set(["flatRoutes", "createRoutesFromFolders"]);
  const result = { sourceFile };
  const visit = (node) => {
    if (result.routesDirArg || result.optionsArg) {
      return;
    }
    if (import_typescript2.default.isCallExpression(node)) {
      const name = extractCalleeName(node.expression);
      if (name && legacyCalls.has(name)) {
        if (name === "flatRoutes") {
          result.routesDirArg = node.arguments[0];
          result.optionsArg = node.arguments[2];
        } else if (name === "createRoutesFromFolders") {
          result.optionsArg = node.arguments[1];
        }
        return;
      }
    }
    import_typescript2.default.forEachChild(node, visit);
  };
  visit(sourceFile);
  return result;
}
function createAutoRoutesOptionsText(printer, legacy) {
  const { routesDirArg, optionsArg, sourceFile } = legacy;
  const hasRoutesDirProperty = optionsArg && import_typescript2.default.isObjectLiteralExpression(optionsArg) && optionsArg.properties.some(isRoutesDirProperty);
  const parts = [];
  if (routesDirArg && !hasRoutesDirProperty) {
    parts.push(
      `routesDir: ${printExpression(printer, sourceFile, routesDirArg)}`
    );
  }
  if (optionsArg) {
    const spreadExpression = printExpression(printer, sourceFile, optionsArg);
    parts.push(`...${spreadExpression}`);
  }
  if (parts.length === 0) {
    return void 0;
  }
  const multiline = parts.length > 1 || parts.some((part) => part.includes("\n"));
  if (!multiline) {
    return `{ ${parts[0]} }`;
  }
  const newline = "\n";
  const body = parts.map((part) => `  ${part},`).join(newline);
  return `{${newline}${body}${newline}}`;
}
function printExpression(printer, sourceFile, expression) {
  return printer.printNode(import_typescript2.default.EmitHint.Expression, expression, sourceFile);
}
function isRoutesDirProperty(property) {
  if (import_typescript2.default.isPropertyAssignment(property) || import_typescript2.default.isShorthandPropertyAssignment(property)) {
    const name = property.name;
    const text = getPropertyNameText(name);
    return text === "routesDir";
  }
  return false;
}
function getPropertyNameText(name) {
  if (import_typescript2.default.isIdentifier(name) || import_typescript2.default.isStringLiteral(name)) {
    return name.text;
  }
  if (import_typescript2.default.isNoSubstitutionTemplateLiteral(name)) {
    return name.text;
  }
  return void 0;
}
function getScriptKind(entryPath) {
  if (entryPath.endsWith(".tsx")) return import_typescript2.default.ScriptKind.TSX;
  if (entryPath.endsWith(".ts")) return import_typescript2.default.ScriptKind.TS;
  if (entryPath.endsWith(".jsx")) return import_typescript2.default.ScriptKind.JSX;
  return import_typescript2.default.ScriptKind.JS;
}
function extractCalleeName(expression) {
  if (import_typescript2.default.isIdentifier(expression)) {
    return expression.text;
  }
  if (import_typescript2.default.isPropertyAccessExpression(expression)) {
    return expression.name.text;
  }
  return null;
}

// src/migration/cli/runner.ts
var import_node_child_process = require("child_process");
var import_node_fs6 = __toESM(require("fs"), 1);
var import_node_os = __toESM(require("os"), 1);
var import_node_path7 = __toESM(require("path"), 1);
var TEMP_PREFIX = "react-router-auto-routes-";
function defaultRunner() {
  const tempDir = import_node_fs6.default.mkdtempSync(import_node_path7.default.join(import_node_os.default.tmpdir(), TEMP_PREFIX));
  const stdoutPath = import_node_path7.default.join(tempDir, "stdout.txt");
  let stdoutFd = null;
  let result = null;
  let extraError;
  try {
    stdoutFd = import_node_fs6.default.openSync(stdoutPath, "w");
    result = (0, import_node_child_process.spawnSync)("npx", ["react-router", "routes"], {
      encoding: "utf8",
      maxBuffer: 50 * 1024 * 1024,
      // 50MB buffer for stderr
      stdio: ["ignore", stdoutFd, "pipe"]
    });
  } catch (error) {
    extraError = error instanceof Error ? error : new Error(String(error));
  } finally {
    if (stdoutFd !== null) {
      try {
        import_node_fs6.default.closeSync(stdoutFd);
      } catch (closeError) {
        if (!extraError && closeError instanceof Error) {
          extraError = closeError;
        }
      }
    }
  }
  let stdout = "";
  try {
    if (import_node_fs6.default.existsSync(stdoutPath)) {
      stdout = import_node_fs6.default.readFileSync(stdoutPath, "utf8");
    }
  } catch (readError) {
    if (!extraError && readError instanceof Error) {
      extraError = readError;
    }
  } finally {
    try {
      import_node_fs6.default.rmSync(tempDir, { recursive: true, force: true });
    } catch (cleanupError) {
      if (!extraError && cleanupError instanceof Error) {
        extraError = cleanupError;
      }
    }
  }
  if (result === null) {
    return {
      status: 1,
      stdout,
      stderr: "",
      error: extraError
    };
  }
  const stderr = (() => {
    const value = result.stderr;
    if (value === void 0 || value === null) {
      return "";
    }
    return typeof value === "string" ? value : value.toString("utf8");
  })();
  return {
    status: result.status ?? 1,
    stdout,
    stderr,
    error: extraError ?? result.error ?? void 0
  };
}
function captureRoutesSnapshot(runner, label) {
  logInfo(`\u25B6\uFE0F  Running "npx react-router routes" (${label})`);
  const result = runner();
  if (result.error) {
    logError('Failed to run "npx react-router routes".');
    logError(result.error);
    return null;
  }
  if (result.status !== 0) {
    logError('"npx react-router routes" exited with a non-zero status.');
    if (result.stderr) {
      logError(result.stderr.trim());
    }
    return null;
  }
  if (result.stderr) {
    const trimmed = result.stderr.trim();
    if (trimmed) {
      logWarn(trimmed);
    }
  }
  return result.stdout ?? "";
}

// src/migration/cli/run-cli.ts
var defaultSourceDir = "app/routes";
function resolveCliArguments(argv) {
  if (argv.length > 2) {
    return null;
  }
  const sourceDir = argv[0] ?? defaultSourceDir;
  const targetDir = argv[1] ?? defaultTargetDir(sourceDir);
  return { sourceDir, targetDir };
}
function buildMigrationPaths(args) {
  const resolvedSource = import_node_path8.default.resolve(args.sourceDir);
  const resolvedTarget = import_node_path8.default.resolve(args.targetDir);
  const parentDir = import_node_path8.default.dirname(resolvedSource);
  const resolvedBackup = import_node_path8.default.join(parentDir, "old-routes");
  return {
    sourceArg: args.sourceDir,
    targetArg: args.targetDir,
    resolvedSource,
    resolvedTarget,
    resolvedBackup,
    parentDir
  };
}
function validateMigrationPaths(paths) {
  const {
    sourceArg,
    targetArg,
    resolvedSource,
    resolvedTarget,
    resolvedBackup
  } = paths;
  const checks = [
    [
      sourceArg === targetArg,
      () => "source and target directories must be different"
    ],
    [
      !import_node_fs7.default.existsSync(sourceArg),
      () => `source directory '${sourceArg}' does not exist`
    ],
    [
      resolvedSource === resolvedBackup,
      () => "source directory cannot be named old-routes"
    ],
    [
      resolvedTarget === resolvedBackup,
      () => "target directory cannot be the backup directory old-routes"
    ],
    [
      import_node_fs7.default.existsSync(resolvedBackup),
      () => `backup directory '${pathRelative(process.cwd(), resolvedBackup)}' already exists. Remove or rename it before running the migration.`
    ]
  ];
  for (const [condition, message] of checks) {
    if (condition) {
      logError(message());
      return false;
    }
  }
  return true;
}
function runCli(argv, options = {}) {
  const args = resolveCliArguments(argv);
  if (!args) {
    usage();
    return 1;
  }
  const paths = buildMigrationPaths(args);
  if (!validateMigrationPaths(paths)) {
    return 1;
  }
  const runner = options.runner ?? defaultRunner;
  const optionsForMigrate = { force: true };
  const context = {
    ...paths,
    runner,
    migrateOptions: optionsForMigrate,
    dryRun: options.dryRun ?? false
  };
  if (!ensureCleanGitWorktree(context.resolvedSource)) {
    return 1;
  }
  return runMigrationWorkflow(context);
}
function runMigrationWorkflow(context) {
  const {
    sourceArg,
    targetArg,
    resolvedSource,
    resolvedTarget,
    resolvedBackup,
    runner,
    migrateOptions,
    dryRun
  } = context;
  const { entryPath, isLegacy } = detectLegacyRouteEntry(resolvedSource);
  let entryRewrite = null;
  const beforeSnapshot = captureRoutesSnapshot(runner, "before migration");
  if (beforeSnapshot === null) {
    return 1;
  }
  try {
    if (import_node_fs7.default.existsSync(resolvedTarget)) {
      import_node_fs7.default.rmSync(resolvedTarget, { recursive: true, force: true });
    }
    migrate(sourceArg, targetArg, migrateOptions);
  } catch (error) {
    logError(error);
    return 1;
  }
  if (dryRun) {
    logInfo(
      `\u{1F9EA} Dry run complete. Review generated routes in '${pathRelative(process.cwd(), resolvedTarget)}'.`
    );
    return 0;
  }
  let swapped = false;
  try {
    swapRoutes(resolvedSource, resolvedTarget, resolvedBackup);
    swapped = true;
    if (isLegacy && entryPath) {
      const rewriteResult = rewriteLegacyRouteEntry(entryPath);
      entryRewrite = rewriteResult.updated ? rewriteResult : null;
      const entryRelative = pathRelative(process.cwd(), entryPath);
      logInfo(
        rewriteResult.updated ? `\u270F\uFE0F Updated route entry '${entryRelative}' to use autoRoutes().` : `\u270F\uFE0F Route entry '${entryRelative}' already references autoRoutes().`
      );
    }
    const afterSnapshot = captureRoutesSnapshot(runner, "after migration");
    if (afterSnapshot === null) {
      entryRewrite = restoreLegacyEntry(entryPath, entryRewrite);
      revertMigration(context, swapped);
      return 1;
    }
    const beforeNormalized = normalizeSnapshot(beforeSnapshot);
    const afterNormalized = normalizeSnapshot(afterSnapshot);
    if (beforeNormalized === afterNormalized) {
      finalizeSuccessfulMigration(resolvedBackup);
      return 0;
    }
    logError("\u274C Route output changed. Reverting migration.");
    logError(diffSnapshots(beforeNormalized, afterNormalized));
    entryRewrite = restoreLegacyEntry(entryPath, entryRewrite);
    revertMigration(context, swapped);
    return 1;
  } catch (error) {
    entryRewrite = restoreLegacyEntry(entryPath, entryRewrite);
    revertMigration(context, swapped);
    logError(error);
    return 1;
  }
}
function restoreLegacyEntry(entryPath, rewrite) {
  if (!rewrite?.updated || !entryPath || typeof rewrite.previousContents !== "string") {
    return null;
  }
  try {
    import_node_fs7.default.writeFileSync(entryPath, rewrite.previousContents);
  } catch (restoreError) {
    const entryRelative = pathRelative(process.cwd(), entryPath);
    logWarn(
      `\u26A0\uFE0F Failed to restore route entry '${entryRelative}'. Restore it manually if needed.`
    );
    logWarn(String(restoreError));
  }
  return null;
}
function revertMigration(context, swapped) {
  if (swapped) {
    revertRoutes(
      context.resolvedSource,
      context.resolvedTarget,
      context.resolvedBackup
    );
    return;
  }
  if (import_node_fs7.default.existsSync(context.resolvedTarget)) {
    import_node_fs7.default.rmSync(context.resolvedTarget, { recursive: true, force: true });
  }
}
function finalizeSuccessfulMigration(resolvedBackup) {
  logInfo("\u2705 Routes match between runs. Migration looks good!");
  if (!import_node_fs7.default.existsSync(resolvedBackup)) {
    return;
  }
  import_node_fs7.default.rmSync(resolvedBackup, { recursive: true, force: true });
  logInfo(
    `\u{1F9F9} Removed temporary backup '${pathRelative(process.cwd(), resolvedBackup)}'. Use git history if you need the previous structure.`
  );
}
function ensureCleanGitWorktree(resolvedSource) {
  const repoCheck = (0, import_node_child_process2.spawnSync)("git", ["rev-parse", "--is-inside-work-tree"], {
    cwd: process.cwd(),
    encoding: "utf8"
  });
  if (repoCheck.status !== 0 || repoCheck.stdout.trim() !== "true") {
    logError(
      "Git repository not detected. Initialize git (or run inside a repository) before using migrate-auto-routes."
    );
    return false;
  }
  const topLevelResult = (0, import_node_child_process2.spawnSync)("git", ["rev-parse", "--show-toplevel"], {
    cwd: process.cwd(),
    encoding: "utf8"
  });
  if (topLevelResult.status !== 0) {
    logError(
      "Unable to determine git repository root. Ensure git is installed and accessible."
    );
    return false;
  }
  const repoRoot = topLevelResult.stdout.trim();
  const sourceRelative = import_node_path8.default.relative(repoRoot, resolvedSource);
  if (sourceRelative.startsWith("..")) {
    logError("Source directory must be inside the git repository.");
    return false;
  }
  const status = (0, import_node_child_process2.spawnSync)("git", ["status", "--porcelain"], {
    cwd: process.cwd(),
    encoding: "utf8"
  });
  if (status.status !== 0) {
    logError(
      "Unable to read git status. Ensure git is installed and accessible."
    );
    return false;
  }
  const dirtyEntries = status.stdout.split("\n").map((line) => line.trim()).filter(Boolean).filter((line) => isWithinSources(line, sourceRelative));
  if (dirtyEntries.length > 0) {
    logError(
      `Working tree must be clean for '${sourceRelative}'. Commit or stash changes affecting this directory and rerun.`
    );
    return false;
  }
  return true;
}
function isWithinSources(statusLine, sourceRelative) {
  if (!sourceRelative || sourceRelative === ".") {
    return statusLine.length > 0;
  }
  const indicator = statusLine.slice(0, 3);
  let remainder = statusLine.slice(3);
  if (indicator.startsWith("R")) {
    const parts = remainder.split(" -> ");
    return parts.some((part) => matchesSource(part, sourceRelative));
  }
  return matchesSource(remainder, sourceRelative);
}
function matchesSource(pathname, sourceRelative) {
  const normalized = pathname.replace(/"/g, "").trim();
  if (normalized === sourceRelative) {
    return true;
  }
  return normalized.startsWith(`${sourceRelative}/`);
}
function usage() {
  logInfo(
    'Usage: migrate-auto-routes [sourceDir] [targetDir]\n\nThe CLI overwrites the target directory if it exists.\n\nThe CLI rewrites routes using the folder + colocation convention promoted by\nreact-router-auto-routes.\n\nWhen sourceDir is omitted, it defaults to "app/routes".\nWhen targetDir is omitted, it defaults to a sibling named "new-routes".'
  );
}

// src/migration/cli.ts
var exitCode = runCli(process.argv.slice(2));
if (exitCode !== 0) {
  process.exit(exitCode);
}
