// react-router-auto-routes/dist/index.js
// src/utils.ts
import * as path2 from "path";

// src/fs/visit-files.ts
import fs from "fs";
import path from "path";
function visitFiles(dir, visitor, options = {}) {
  const { followSymlinks = true } = options;
  const rootDir = dir;
  const readStat = followSymlinks ? fs.statSync : fs.lstatSync;
  const walk = (currentDir) => {
    for (const filename of fs.readdirSync(currentDir)) {
      const file = path.resolve(currentDir, filename);
      const stat = readStat(file);
      if (stat.isDirectory()) {
        walk(file);
      } else if (stat.isFile()) {
        visitor(path.relative(rootDir, file));
      }
    }
  };
  walk(dir);
}

// src/utils.ts
function defaultVisitFiles(dir, visitor) {
  visitFiles(dir, visitor);
}
function createRouteId(file) {
  const normalized = file.split(path2.win32.sep).join("/");
  return normalized.replace(/\.[a-z0-9]+$/i, "");
}
function normalizeRoutesDir(routeDir) {
  const trimmed = routeDir.trim();
  if (!trimmed) {
    throw new Error(
      `routesDir entries must be non-empty strings. Got: '${routeDir}'`
    );
  }
  let normalized = trimmed.replace(/\\/g, "/");
  normalized = normalized.replace(/\/+/g, "/");
  while (normalized.endsWith("/")) {
    normalized = normalized.slice(0, -1);
  }
  if (!normalized) {
    throw new Error(
      `routesDir entries must resolve to a folder. Got: '${routeDir}'`
    );
  }
  if (normalized.startsWith("/")) {
    throw new Error(
      `routesDir entries must be relative paths, not absolute. Got: '${routeDir}'`
    );
  }
  const segments = normalized.split("/");
  if (segments.some((segment) => segment === "" || segment === ".")) {
    throw new Error(
      `routesDir entries cannot contain '.' segments. Got: '${routeDir}'`
    );
  }
  return normalized;
}
function normalizeMountPath(mountPath) {
  if (!mountPath.trim()) {
    throw new Error("routesDir mount paths must be non-empty strings.");
  }
  const normalized = mountPath.trim();
  if (!normalized.startsWith("/")) {
    throw new Error(
      `routesDir mount paths must start with '/'. Got: '${mountPath}'`
    );
  }
  if (normalized !== "/" && normalized.endsWith("/")) {
    throw new Error(
      `routesDir mount paths cannot end with '/'. Got: '${mountPath}'`
    );
  }
  return normalized;
}
function toRoutesDirEntries(routesDir) {
  if (routesDir === void 0) {
    return [{ mountPath: "/", dir: "routes" }];
  }
  if (typeof routesDir === "string") {
    return [{ mountPath: "/", dir: routesDir }];
  }
  const entries = [];
  for (const [mountPath, dir] of Object.entries(routesDir)) {
    entries.push({ mountPath, dir });
  }
  if (entries.length === 0) {
    throw new Error("routesDir object must contain at least one entry.");
  }
  return entries;
}
function getRootDirectory() {
  return process.cwd();
}
function getAppDirectory() {
  return globalThis.__reactRouterAppDirectory ?? path2.resolve(getRootDirectory(), "app");
}
function resolveBaseDirFor(routesDir) {
  return routesDir === void 0 || typeof routesDir === "string" ? getAppDirectory() : getRootDirectory();
}
function normalizeRoutesDirOption(routesDir) {
  const resolvedBase = resolveBaseDirFor(routesDir);
  const seenMountPaths = /* @__PURE__ */ new Set();
  const entries = toRoutesDirEntries(routesDir).map(({ mountPath, dir }) => {
    const normalizedMount = normalizeMountPath(mountPath);
    if (seenMountPaths.has(normalizedMount)) {
      throw new Error(
        `Duplicate routesDir mount path detected: '${normalizedMount}'.`
      );
    }
    seenMountPaths.add(normalizedMount);
    const normalizedDir = normalizeRoutesDir(dir);
    const fsDir = path2.resolve(resolvedBase, normalizedDir);
    const idPrefix = normalizedMount === "/" && normalizedDir === "app/routes" ? "routes" : normalizedDir;
    return {
      mountPath: normalizedMount,
      fsDir,
      idPrefix,
      importPrefix: normalizedDir
    };
  });
  const rootEntry = entries.find((entry) => entry.mountPath === "/");
  const appDir = rootEntry ? path2.dirname(rootEntry.fsDir) : getAppDirectory();
  return entries.map((entry) => {
    const relativeImport = path2.relative(appDir, entry.fsDir) || ".";
    const normalizedImport = relativeImport.split(path2.sep).join("/");
    return {
      ...entry,
      importPrefix: normalizedImport === "." ? "" : normalizedImport
    };
  });
}

// src/core/routing/files.ts
import * as path4 from "path";
import picomatch from "picomatch";

// src/core/routing/constants.ts
var ROUTE_EXTENSIONS = [".js", ".jsx", ".ts", ".tsx", ".md", ".mdx"];
var SPECIAL_ROUTE_FILES = [
  "index",
  "_index",
  "route",
  "_route",
  "_layout"
];
var SERVER_FILE_REGEX = /\.server\.(ts|tsx|js|jsx|md|mdx)$/;
var ROOT_PARENT = "root";

// src/core/routing/segments.ts
import path3 from "path";
var ESCAPE_START = "[";
var ESCAPE_END = "]";
var OPTIONAL_START = "(";
var OPTIONAL_END = ")";
var WINDOWS_SEPARATOR = path3.win32.sep;
var STRIP_ALWAYS = /* @__PURE__ */ new Set(["route", "_route"]);
function isSegmentSeparator(char) {
  if (!char) return false;
  return char === "/" || char === "." || char === WINDOWS_SEPARATOR;
}
function pushRouteSegment(routeSegments, rawRouteSegments, segment, rawSegment, routeId) {
  if (!segment) return;
  const checks = [
    [rawSegment, "*", rawSegment],
    [rawSegment, ":", rawSegment],
    [rawSegment, "/", segment]
  ];
  for (const [testValue, char, value] of checks) {
    if (testValue.includes(char)) {
      throw new Error(
        `Route segment "${value}" for "${routeId}" cannot contain "${char}".
If this is something you need, upvote this proposal for React Router https://github.com/remix-run/react-router/discussions/9822.`
      );
    }
  }
  routeSegments.push(segment);
  rawRouteSegments.push(rawSegment);
}
function stripTrailingSpecialSegments(originalName, routeSegments, rawRouteSegments) {
  if (!originalName.includes("/")) {
    return;
  }
  while (routeSegments.length > 0 && rawRouteSegments.length > 0) {
    const lastRaw = rawRouteSegments[rawRouteSegments.length - 1];
    if (!lastRaw || !STRIP_ALWAYS.has(lastRaw)) {
      break;
    }
    routeSegments.pop();
    rawRouteSegments.pop();
  }
}
function getRouteSegments(routeId, isIndex, paramChar = "$") {
  const routeSegments = [];
  const rawRouteSegments = [];
  let routeSegment = "";
  let rawRouteSegment = "";
  let state = "NORMAL";
  const flushSegment = () => {
    pushRouteSegment(
      routeSegments,
      rawRouteSegments,
      routeSegment,
      rawRouteSegment,
      routeId
    );
    routeSegment = "";
    rawRouteSegment = "";
  };
  for (let index = 0; index < routeId.length; index++) {
    const char = routeId[index];
    if (state === "ESCAPE" || state === "OPTIONAL_ESCAPE") {
      if (char === ESCAPE_END) {
        state = state === "ESCAPE" ? "NORMAL" : "OPTIONAL";
        rawRouteSegment += char;
      } else {
        routeSegment += char;
        rawRouteSegment += char;
      }
      continue;
    }
    if (state === "NORMAL") {
      if (isSegmentSeparator(char)) {
        flushSegment();
        continue;
      }
      if (char === ESCAPE_START) {
        state = "ESCAPE";
        rawRouteSegment += char;
        continue;
      }
      if (char === OPTIONAL_START) {
        state = "OPTIONAL";
        rawRouteSegment += char;
        continue;
      }
    }
    if (state === "OPTIONAL") {
      if (char === OPTIONAL_END) {
        routeSegment += "?";
        rawRouteSegment += char;
        state = "NORMAL";
        continue;
      }
      if (char === ESCAPE_START) {
        state = "OPTIONAL_ESCAPE";
        rawRouteSegment += char;
        continue;
      }
    }
    if (!routeSegment && char === paramChar && (state === "NORMAL" || state === "OPTIONAL")) {
      const next = routeId[index + 1];
      routeSegment += index + 1 === routeId.length || isSegmentSeparator(next) ? "*" : ":";
      rawRouteSegment += char;
      continue;
    }
    routeSegment += char;
    rawRouteSegment += char;
  }
  flushSegment();
  stripTrailingSpecialSegments(routeId, routeSegments, rawRouteSegments);
  const path5 = createRoutePath(routeSegments, rawRouteSegments, isIndex);
  const segments = deriveNameSegments(routeId, rawRouteSegments);
  return { segments, path: path5 };
}
function createRoutePath(routeSegments, rawRouteSegments, isIndex) {
  const length = isIndex ? routeSegments.length - 1 : routeSegments.length;
  if (length <= 0) {
    return void 0;
  }
  const parts = [];
  for (let i = 0; i < length; i++) {
    let segment = routeSegments[i];
    const rawSegment = rawRouteSegments[i];
    if (!rawSegment) {
      continue;
    }
    if (segment.startsWith("_") && rawSegment.startsWith("_")) {
      continue;
    }
    if (segment.endsWith("_") && rawSegment.endsWith("_")) {
      segment = segment.slice(0, -1);
    }
    parts.push(segment);
  }
  return parts.length > 0 ? parts.join("/") : void 0;
}
function deriveNameSegments(routeId, rawSegments) {
  const segments = [...rawSegments];
  while (segments.length > 0) {
    const last = segments[segments.length - 1];
    if (!last || !SPECIAL_ROUTE_FILES.includes(last)) {
      break;
    }
    segments.pop();
  }
  return segments;
}

// src/core/routing/utils.ts
var regexCache = {};
function memoizedRegex(input) {
  if (input in regexCache) {
    return regexCache[input];
  }
  const newRegex = new RegExp(input);
  regexCache[input] = newRegex;
  return newRegex;
}
function escapeRegexChar(char) {
  return char.replace(/[.*+\-?^${}()|[\]\\]/g, "\\$&");
}
function resolveRouteRegex(pattern, colocationChar) {
  const escapedColocationChar = escapeRegexChar(colocationChar);
  return new RegExp(
    pattern.source.replace("\\${colocationChar}", `\\${escapedColocationChar}`),
    pattern.flags
  );
}

// src/core/routing/files.ts
var routeModuleExts = ROUTE_EXTENSIONS;
var serverRegex = SERVER_FILE_REGEX;
function checkColocationViolations(segments, colocationChar, filename) {
  if (segments.length >= 1 && segments[0].startsWith(colocationChar)) {
    throw new Error(
      `Colocation entries must live inside a route folder. Move '${filename}' under an actual route directory.`
    );
  }
  let anonymousFolderCount = 0;
  for (let i = 0; i < segments.length - 1; i++) {
    if (segments[i] === colocationChar) {
      anonymousFolderCount++;
      if (anonymousFolderCount > 1) {
        throw new Error(
          `Nested anonymous colocation folders (+/+/) are not allowed. Use named folders like +/components/ instead. Found in: ${filename}`
        );
      }
    }
  }
}
function isColocated(segments, colocationChar) {
  return segments.some((segment) => segment.startsWith(colocationChar));
}
function hasValidRouteExtension(filename, routeRegex) {
  const isFlatFile = !filename.includes(path4.sep) && !filename.includes("/");
  if (serverRegex.test(filename)) {
    return false;
  }
  if (isFlatFile) {
    return routeModuleExts.includes(path4.extname(filename));
  }
  if (routeRegex) {
    return routeRegex.test(filename);
  }
  return routeModuleExts.includes(path4.extname(filename));
}
function isRouteModuleFile(filename, colocationChar = "+", routeRegex) {
  const normalizedPath = filename.replace(/\\/g, "/");
  const segments = normalizedPath.split("/");
  checkColocationViolations(segments, colocationChar, filename);
  if (isColocated(segments, colocationChar)) {
    return false;
  }
  return hasValidRouteExtension(filename, routeRegex);
}
function isIndexRoute(routeId) {
  const indexRouteRegex = memoizedRegex(
    `((^|[.])(index|_index))($|\\/)|(\\/(_?index))($|\\/)`
  );
  return indexRouteRegex.test(routeId);
}
function applyMountPath(routePath, mountPath) {
  if (mountPath === "/") {
    return routePath;
  }
  const normalizedMount = mountPath.endsWith("/") ? mountPath.slice(0, -1) : mountPath;
  if (!routePath) {
    return normalizedMount;
  }
  const normalizedPath = routePath.startsWith("/") ? routePath : `/${routePath}`;
  return `${normalizedMount}${normalizedPath}`;
}
function getRouteInfo(routeDir, file, options) {
  const idFilePath = path4.join(routeDir.idPrefix, file).split(path4.win32.sep).join("/");
  const importFilePath = path4.join(routeDir.importPrefix, file).split(path4.win32.sep).join("/");
  const routeId = createRouteId(idFilePath);
  const routePrefix = `${routeDir.idPrefix}/`;
  if (!routeId.startsWith(routePrefix)) {
    throw new Error(
      `Route id '${routeId}' does not start with expected prefix '${routePrefix}'`
    );
  }
  const routeIdWithoutRoutes = routeId.slice(routePrefix.length);
  const index = isIndexRoute(routeIdWithoutRoutes);
  const { segments, path: routePath } = getRouteSegments(
    routeIdWithoutRoutes,
    index,
    options.paramChar
  );
  const pathWithMount = applyMountPath(routePath, routeDir.mountPath);
  return {
    id: routeId,
    path: pathWithMount,
    file: importFilePath,
    segments,
    mountPath: routeDir.mountPath,
    index
  };
}
function collectRouteInfos(options) {
  const { routes, ignoredRouteFiles, visitFiles: visitFiles2, colocationChar, routeRegex } = options;
  const routeMap = /* @__PURE__ */ new Map();
  for (const currentRouteDir of routes) {
    visitFiles2(currentRouteDir.fsDir, (file) => {
      if (ignoredRouteFiles.length > 0 && ignoredRouteFiles.some(
        (pattern) => picomatch.isMatch(file, pattern, { dot: true })
      )) {
        return;
      }
      if (!isRouteModuleFile(file, colocationChar, routeRegex)) {
        return;
      }
      const routeInfo = getRouteInfo(currentRouteDir, file, options);
      const existing = routeMap.get(routeInfo.id);
      if (existing) {
        throw new Error(
          `Duplicate route id '${routeInfo.id}' generated from '${routeInfo.file}' conflicts with '${existing.file}'`
        );
      }
      routeMap.set(routeInfo.id, routeInfo);
    });
  }
  return Array.from(routeMap.values());
}

// src/core/routing/structure.ts
function isLayoutRoute(route) {
  const fileName = route.file.split("/").pop() ?? "";
  const normalized = fileName.replace(/\.[^/.]+$/, "");
  return normalized === "_layout" || normalized.endsWith("._layout");
}
function isParentCandidate(route) {
  return isLayoutRoute(route);
}
var ROUTE_PRIORITY = 2;
var LAYOUT_PRIORITY = 3;
var ROUTE_DIR_PRIORITY = 4;
function parentScore(route) {
  const normalizedFile = route.file.replace(/\\/g, "/");
  if (normalizedFile.includes("/route/route.")) {
    return ROUTE_DIR_PRIORITY;
  }
  if (isLayoutRoute(route)) {
    return LAYOUT_PRIORITY;
  }
  return ROUTE_PRIORITY;
}
function keyFor(mountPath, segments) {
  return `${mountPath}::${segments.join("/")}`;
}
function resolveRelativePath(route, parentPath) {
  if (!route.path) {
    return void 0;
  }
  let relative2 = route.path;
  if (parentPath && relative2.startsWith(parentPath)) {
    relative2 = relative2.slice(parentPath.length);
    if (relative2.startsWith("/")) {
      relative2 = relative2.slice(1);
    }
  } else if (relative2.startsWith("/")) {
    relative2 = relative2.slice(1);
  }
  if (route.index && parentPath && !relative2) {
    return void 0;
  }
  return relative2 || void 0;
}
function segmentSpecificity(segment) {
  if (segment === "*") {
    return 0;
  }
  if (segment.startsWith(":")) {
    return segment.endsWith("?") ? 1 : 2;
  }
  return segment.endsWith("?") ? 3 : 4;
}
function comparePathSpecificity(a, b) {
  if (!a.path || !b.path) {
    return 0;
  }
  const left = a.path.split("/").filter(Boolean);
  const right = b.path.split("/").filter(Boolean);
  const length = Math.min(left.length, right.length);
  for (let index = 0; index < length; index++) {
    const specificityDiff = segmentSpecificity(right[index]) - segmentSpecificity(left[index]);
    if (specificityDiff !== 0) {
      return specificityDiff;
    }
  }
  return 0;
}
function sortForAssembly(routes) {
  return [...routes].sort((a, b) => {
    const depthDiff = a.segments.length - b.segments.length;
    if (depthDiff !== 0) {
      return depthDiff;
    }
    if (a.index !== b.index) {
      return a.index ? 1 : -1;
    }
    const scoreDiff = parentScore(b) - parentScore(a);
    if (scoreDiff !== 0) {
      return scoreDiff;
    }
    const specificityDiff = comparePathSpecificity(a, b);
    if (specificityDiff !== 0) {
      return specificityDiff;
    }
    return a.id.localeCompare(b.id);
  });
}
function buildRouteTree(routes) {
  const sortedRoutes = sortForAssembly(routes);
  const configMap = /* @__PURE__ */ new Map();
  const absolutePaths = /* @__PURE__ */ new Map();
  const parentLookup = /* @__PURE__ */ new Map();
  const roots = [];
  const resolveParentId = (route) => {
    let parentId;
    if (route.index) {
      parentId = parentLookup.get(keyFor(route.mountPath, route.segments))?.id;
    }
    if (!parentId) {
      for (let length = route.segments.length - 1; length >= 0; length--) {
        const candidate = parentLookup.get(
          keyFor(route.mountPath, route.segments.slice(0, length))
        );
        if (candidate && candidate.id !== route.id) {
          parentId = candidate.id;
          break;
        }
      }
    }
    if (!parentId && route.segments.length === 0) {
      const candidate = parentLookup.get(keyFor(route.mountPath, []));
      if (candidate && candidate.id !== route.id) {
        parentId = candidate.id;
      }
    }
    return parentId ?? ROOT_PARENT;
  };
  for (const route of sortedRoutes) {
    const parentId = resolveParentId(route);
    const parentPath = parentId === ROOT_PARENT ? "/" : absolutePaths.get(parentId);
    const config = {
      id: route.id,
      file: route.file
    };
    const relativePath = resolveRelativePath(
      route,
      parentPath === "/" ? void 0 : parentPath
    );
    if (relativePath !== void 0) {
      config.path = relativePath;
    }
    if (route.index) {
      config.index = true;
    }
    if (!route.index && isParentCandidate(route)) {
      const key = keyFor(route.mountPath, route.segments);
      const current = parentLookup.get(key);
      const score = parentScore(route);
      if (!current) {
        parentLookup.set(key, route);
      } else {
        const currentScore = parentScore(current);
        if (score > currentScore || score === currentScore && route.id < current.id) {
          parentLookup.set(key, route);
        }
      }
    }
    if (parentId === ROOT_PARENT) {
      roots.push(config);
      configMap.set(route.id, config);
      absolutePaths.set(route.id, route.path);
      continue;
    }
    const parentConfig = configMap.get(parentId);
    if (!parentConfig) {
      roots.push(config);
      configMap.set(route.id, config);
      absolutePaths.set(route.id, route.path);
      continue;
    }
    if (parentConfig.index) {
      throw new Error(
        `Child routes are not allowed in index routes. Please remove child routes of ${parentConfig.id}`
      );
    }
    if (!parentConfig.children) {
      parentConfig.children = [];
    }
    parentConfig.children.push(config);
    configMap.set(route.id, config);
    absolutePaths.set(route.id, route.path);
  }
  return roots;
}

// src/core/index.ts
var DEFAULT_ROUTE_REGEX = /((\${colocationChar}[\/\\][^\/\\:?*]+)|[\/\\]((index|route)|(_[^\/\\:?*]+)|([^\/\\:?*]+\.route)|([^\/\\:?*]+)))\.(ts|tsx|js|jsx|md|mdx)$/;
function autoRoutes(options = {}) {
  const {
    routesDir,
    paramChar = "$",
    colocationChar: userColocationChar,
    routeRegex: userRouteRegex,
    ignoredRouteFiles = [],
    visitFiles: userVisitFiles
  } = options;
  const colocationChar = userColocationChar ?? "+";
  const visitFiles2 = userVisitFiles ?? defaultVisitFiles;
  const routeRegex = resolveRouteRegex(
    userRouteRegex ?? DEFAULT_ROUTE_REGEX,
    colocationChar
  );
  const routes = normalizeRoutesDirOption(routesDir);
  const resolved = {
    routes,
    visitFiles: visitFiles2,
    paramChar,
    colocationChar,
    ignoredRouteFiles: [...ignoredRouteFiles],
    routeRegex
  };
  const collectedRoutes = collectRouteInfos(resolved);
  return buildRouteTree(collectedRoutes);
}
export {
  autoRoutes,
  autoRoutes as default
};
