// @conform-to/react/dist/future/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var future = require('@conform-to/dom/future');
var forms = require('./forms.js');
var hooks = require('./hooks.js');
var util = require('./util.js');
var memoize = require('./memoize.js');



Object.defineProperty(exports, 'getFieldValue', {
	enumerable: true,
	get: function () { return future.getFieldValue; }
});
Object.defineProperty(exports, 'isDirty', {
	enumerable: true,
	get: function () { return future.isDirty; }
});
Object.defineProperty(exports, 'parseSubmission', {
	enumerable: true,
	get: function () { return future.parseSubmission; }
});
Object.defineProperty(exports, 'report', {
	enumerable: true,
	get: function () { return future.report; }
});
exports.configureForms = forms.configureForms;
exports.BaseControl = hooks.BaseControl;
exports.FormOptionsProvider = hooks.FormOptionsProvider;
exports.FormProvider = hooks.FormProvider;
exports.PreserveBoundary = hooks.PreserveBoundary;
exports.useControl = hooks.useControl;
exports.useField = hooks.useField;
exports.useForm = hooks.useForm;
exports.useFormData = hooks.useFormData;
exports.useFormMetadata = hooks.useFormMetadata;
exports.useIntent = hooks.useIntent;
exports.shape = util.shape;
exports.memoize = memoize.memoize;
