// @conform-to/react/dist/future/intent.mjs
import { objectSpread2 as _objectSpread2 } from '../_virtual/_rollupPluginBabelHelpers.mjs';
import { parsePath, getRelativePath, isPlainObject, appendPath, getPathValue } from '@conform-to/dom/future';
import { isOptional, isUndefined, isNullable, appendUniqueItem, merge, updatePathValue, isString, getPathArray, createPathIndexUpdater, compactMap, generateUniqueKey, isNumber, transformKeys } from './util.mjs';
import { getDefaultListKey } from './state.mjs';

/**
 * Serializes intent to string format: "type" or "type(payload)".
 */
function serializeIntent(intent) {
  if (!intent.payload) {
    return intent.type;
  }
  return "".concat(intent.type, "(").concat(JSON.stringify(intent.payload), ")");
}

/**
 * Parses serialized intent string back to intent object.
 */
function deserializeIntent(value) {
  var type = value;
  var payload;
  var serializedPayload;
  var openParenIndex = value.indexOf('(');
  if (openParenIndex > 0 && value[value.length - 1] === ')') {
    type = value.slice(0, openParenIndex);
    serializedPayload = value.slice(openParenIndex + 1, -1);
  }
  if (serializedPayload) {
    try {
      payload = JSON.parse(serializedPayload);
    } catch (_unused) {
      // Ignore the error
    }
  }
  return {
    type,
    payload
  };
}

/**
 * Applies intent transformation to submission payload.
 * Returns modified payload or null for reset intent.
 */
function resolveIntent(submission, options) {
  var _options$handlers, _handler$validate, _handler$validate2;
  if (!submission.intent) {
    return submission.payload;
  }
  var intent = deserializeIntent(submission.intent);
  var handlers = (_options$handlers = options === null || options === void 0 ? void 0 : options.handlers) !== null && _options$handlers !== void 0 ? _options$handlers : intentHandlers;
  var handler = handlers[intent.type];
  if (handler !== null && handler !== void 0 && handler.resolve && ((_handler$validate = (_handler$validate2 = handler.validate) === null || _handler$validate2 === void 0 ? void 0 : _handler$validate2.call(handler, intent.payload)) !== null && _handler$validate !== void 0 ? _handler$validate : true)) {
    return handler.resolve(submission.payload, intent.payload);
  }
  return submission.payload;
}

/**
 * Resolves an intent after validation by calling the handler's onResolve.
 * Mutates the result with updated value/error and returns whether the intent was cancelled.
 */
function applyIntent(result, intent, options) {
  if (intent) {
    var _options$handlers2, _handler$validate3, _handler$validate4;
    var handlers = (_options$handlers2 = options === null || options === void 0 ? void 0 : options.handlers) !== null && _options$handlers2 !== void 0 ? _options$handlers2 : intentHandlers;
    var handler = handlers[intent.type];
    if (handler !== null && handler !== void 0 && handler.apply && ((_handler$validate3 = (_handler$validate4 = handler.validate) === null || _handler$validate4 === void 0 ? void 0 : _handler$validate4.call(handler, intent.payload)) !== null && _handler$validate3 !== void 0 ? _handler$validate3 : true)) {
      return handler.apply(result, intent.payload);
    }
  }
  return result;
}
function insertItem(list, item, index) {
  list.splice(index, 0, item);
}
function removeItem(list, index) {
  list.splice(index, 1);
}
function reorderItems(list, fromIndex, toIndex) {
  list.splice(toIndex, 0, ...list.splice(fromIndex, 1));
}

/**
 * Updates list keys by removing child keys and optionally transforming remaining keys.
 */
function updateListKeys() {
  var keys = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var keyToBeRemoved = arguments.length > 1 ? arguments[1] : undefined;
  var updateKey = arguments.length > 2 ? arguments[2] : undefined;
  var basePath = parsePath(keyToBeRemoved);
  return transformKeys(keys, field => {
    var _updateKey;
    return getRelativePath(field, basePath) !== null ? null : (_updateKey = updateKey === null || updateKey === void 0 ? void 0 : updateKey(field)) !== null && _updateKey !== void 0 ? _updateKey : field;
  });
}

/**
 * Built-in action handlers for form intents:
 * - reset: clears form data
 * - validate: marks fields as touched for validation display
 * - update: updates specific field values
 * - insert/remove/reorder: manages array field operations
 */
var intentHandlers = {
  reset: {
    validate(options) {
      return isOptional(options, isPlainObject) && (isUndefined(options === null || options === void 0 ? void 0 : options.defaultValue) || isNullable(options === null || options === void 0 ? void 0 : options.defaultValue, isPlainObject));
    },
    resolve(_, options) {
      if ((options === null || options === void 0 ? void 0 : options.defaultValue) === null) {
        return {};
      }
      return options === null || options === void 0 ? void 0 : options.defaultValue;
    },
    apply(result) {
      return _objectSpread2(_objectSpread2({}, result), {}, {
        reset: true
      });
    }
  },
  validate: {
    validate(name) {
      return isOptional(name, isString);
    },
    update(state, _ref) {
      var _intent$payload;
      var {
        submission,
        intent,
        error
      } = _ref;
      var name = (_intent$payload = intent.payload) !== null && _intent$payload !== void 0 ? _intent$payload : '';
      var basePath = parsePath(name);
      var allFields = error ?
      // Consider fields / fieldset with errors as touched too
      submission.fields.concat(Object.keys(error.fieldErrors)) : submission.fields;
      var touchedFields = appendUniqueItem(state.touchedFields, name);
      for (var field of allFields) {
        // Add all child fields to the touched fields too
        if (getRelativePath(field, basePath) !== null) {
          touchedFields = appendUniqueItem(touchedFields, field);
        }
      }
      return merge(state, {
        touchedFields
      });
    }
  },
  update: {
    validate(options) {
      return isPlainObject(options) && isOptional(options.name, isString) && isOptional(options.index, isNumber) && !isUndefined(options.value);
    },
    resolve(value, options) {
      var _options$value;
      var name = appendPath(options.name, options.index);
      return updatePathValue(value, name, (_options$value = options.value) !== null && _options$value !== void 0 ? _options$value : name === '' ? {} : null);
    },
    update(state, _ref2) {
      var {
        type,
        submission,
        intent
      } = _ref2;
      if (type === 'server') {
        return state;
      }
      var listKeys = state.listKeys;

      // Update the keys only for client updates to avoid double updates if there is no client validation
      if (type === 'client') {
        // TODO: Do we really need to update the keys here?
        var name = appendPath(intent.payload.name, intent.payload.index);
        // Remove all child keys
        listKeys = name === '' ? {} : updateListKeys(state.listKeys, name);
      }
      var basePath = parsePath(intent.payload.name);
      var touchedFields = state.touchedFields;
      for (var field of submission.fields) {
        if (basePath.length === 0 || getRelativePath(field, basePath) !== null) {
          touchedFields = appendUniqueItem(touchedFields, field);
        }
      }
      return _objectSpread2(_objectSpread2({}, state), {}, {
        listKeys,
        touchedFields
      });
    }
  },
  insert: {
    validate(options) {
      return isPlainObject(options) && isString(options.name) && isOptional(options.index, isNumber) && isOptional(options.from, isString) && isOptional(options.onInvalid, mode => mode === 'revert');
    },
    resolve(value, options) {
      var _options$index;
      var result = value;
      var itemValue = options.defaultValue;
      if (options.from !== undefined) {
        itemValue = getPathValue(result, options.from);
        result = updatePathValue(result, options.from, '');
      }
      var list = Array.from(getPathArray(result, options.name));
      insertItem(list, itemValue, (_options$index = options.index) !== null && _options$index !== void 0 ? _options$index : list.length);
      return updatePathValue(result, options.name, list);
    },
    apply(result, options) {
      var _result$error;
      // Warn if validation result is not yet available
      if (typeof result.error === 'undefined' && (options.onInvalid || options.from)) {
        // eslint-disable-next-line no-console
        console.warn('intent.insert() with `onInvalid` or `from` requires the validation result to be available synchronously. ' + 'These options are ignored because the error is not yet known.');
        return result;
      }
      var arrayErrors = (_result$error = result.error) === null || _result$error === void 0 ? void 0 : _result$error.fieldErrors[options.name];
      if (options.onInvalid === 'revert' && arrayErrors !== null && arrayErrors !== void 0 && arrayErrors.length) {
        return _objectSpread2(_objectSpread2({}, result), {}, {
          targetValue: undefined
        });
      }
      if (options.from !== undefined) {
        var _options$index2, _result$error2;
        var index = (_options$index2 = options.index) !== null && _options$index2 !== void 0 ? _options$index2 : getPathArray(result.submission.payload, options.name).length;
        var insertedItemPath = appendPath(options.name, index);
        var insertedItemErrors = (_result$error2 = result.error) === null || _result$error2 === void 0 ? void 0 : _result$error2.fieldErrors[insertedItemPath];
        if (insertedItemErrors !== null && insertedItemErrors !== void 0 && insertedItemErrors.length) {
          var _result$error$fieldEr, _result$error3, _result$error$formErr, _result$error4, _result$error5;
          var fromErrors = (_result$error$fieldEr = (_result$error3 = result.error) === null || _result$error3 === void 0 ? void 0 : _result$error3.fieldErrors[options.from]) !== null && _result$error$fieldEr !== void 0 ? _result$error$fieldEr : [];
          return _objectSpread2(_objectSpread2({}, result), {}, {
            targetValue: undefined,
            error: {
              formErrors: (_result$error$formErr = (_result$error4 = result.error) === null || _result$error4 === void 0 ? void 0 : _result$error4.formErrors) !== null && _result$error$formErr !== void 0 ? _result$error$formErr : [],
              fieldErrors: _objectSpread2(_objectSpread2({}, (_result$error5 = result.error) === null || _result$error5 === void 0 ? void 0 : _result$error5.fieldErrors), {}, {
                [options.from]: [...fromErrors, ...insertedItemErrors],
                [insertedItemPath]: []
              })
            }
          });
        }
      }
      return result;
    },
    update(state, _ref3) {
      var _intent$payload$index;
      var {
        type,
        submission,
        intent,
        ctx
      } = _ref3;
      if (type === 'server') {
        return state;
      }
      var from = intent.payload.from;
      var index = (_intent$payload$index = intent.payload.index) !== null && _intent$payload$index !== void 0 ? _intent$payload$index : getPathArray(submission.payload, intent.payload.name).length;
      var updateListIndex = createPathIndexUpdater(intent.payload.name, currentIndex => index <= currentIndex ? currentIndex + 1 : currentIndex);
      var touchedFields = state.touchedFields;
      var listKeys = state.listKeys;
      if (!ctx.cancelled) {
        touchedFields = compactMap(state.touchedFields, updateListIndex);

        // Update the keys only for client updates to avoid double updates if there is no client validation
        if (type === 'client') {
          var _state$listKeys$inten;
          var selectedListKeys = Array.from((_state$listKeys$inten = state.listKeys[intent.payload.name]) !== null && _state$listKeys$inten !== void 0 ? _state$listKeys$inten : getDefaultListKey(state.resetKey, submission.payload, intent.payload.name));
          insertItem(selectedListKeys, generateUniqueKey(), index);
          listKeys = _objectSpread2(_objectSpread2({}, updateListKeys(state.listKeys, appendPath(intent.payload.name, index), updateListIndex)), {}, {
            // Update existing list keys
            [intent.payload.name]: selectedListKeys
          });
        }
      }
      touchedFields = appendUniqueItem(touchedFields, intent.payload.name);
      if (from !== undefined) {
        touchedFields = appendUniqueItem(touchedFields, from);
      }
      return _objectSpread2(_objectSpread2({}, state), {}, {
        listKeys,
        touchedFields
      });
    }
  },
  remove: {
    validate(options) {
      return isPlainObject(options) && isString(options.name) && isNumber(options.index) && isOptional(options.onInvalid, v => v === 'revert' || v === 'insert');
    },
    resolve(value, options) {
      var list = Array.from(getPathArray(value, options.name));
      removeItem(list, options.index);
      return updatePathValue(value, options.name, list);
    },
    apply(result, options) {
      var _result$error6;
      // Warn if validation result is not yet available
      if (typeof result.error === 'undefined' && options.onInvalid) {
        if (process.env.NODE_ENV !== 'production') {
          // eslint-disable-next-line no-console
          console.warn('intent.remove() with `onInvalid` requires the validation result to be available synchronously. ' + 'This option is ignored because the error is not yet known.');
        }
        return result;
      }
      if (result.targetValue && (_result$error6 = result.error) !== null && _result$error6 !== void 0 && _result$error6.fieldErrors[options.name]) {
        switch (options.onInvalid) {
          case 'revert':
            return _objectSpread2(_objectSpread2({}, result), {}, {
              targetValue: undefined
            });
          case 'insert':
            {
              var list = Array.from(getPathArray(result.targetValue, options.name));
              insertItem(list, options.defaultValue, list.length);
              return _objectSpread2(_objectSpread2({}, result), {}, {
                targetValue: updatePathValue(result.targetValue, options.name, list)
              });
            }
        }
      }
      return result;
    },
    update(state, _ref4) {
      var {
        type,
        submission,
        intent,
        ctx
      } = _ref4;
      if (type === 'server') {
        return state;
      }
      var currentValue = submission.payload;
      var updateListIndex = createPathIndexUpdater(intent.payload.name, currentIndex => {
        if (intent.payload.index === currentIndex) {
          return null;
        }
        return intent.payload.index < currentIndex ? currentIndex - 1 : currentIndex;
      });
      var touchedFields = state.touchedFields;
      var listKeys = state.listKeys;

      // If onInvalid is 'insert', we still remove the item and then insert a new item at the end
      if (!ctx.cancelled || intent.payload.onInvalid === 'insert') {
        touchedFields = compactMap(touchedFields, updateListIndex);

        // Update the keys only for client updates to avoid double updates if there is no client validation
        if (type === 'client') {
          var _state$listKeys$inten2;
          var selectedListKeys = Array.from((_state$listKeys$inten2 = state.listKeys[intent.payload.name]) !== null && _state$listKeys$inten2 !== void 0 ? _state$listKeys$inten2 : getDefaultListKey(state.resetKey, currentValue, intent.payload.name));
          removeItem(selectedListKeys, intent.payload.index);
          listKeys = _objectSpread2(_objectSpread2({}, updateListKeys(state.listKeys, appendPath(intent.payload.name, intent.payload.index), updateListIndex)), {}, {
            // Update existing list keys
            [intent.payload.name]: selectedListKeys
          });
          if (ctx.cancelled) {
            var index = selectedListKeys.length;
            insertItem(selectedListKeys, generateUniqueKey(), index);
            listKeys = _objectSpread2(_objectSpread2({}, updateListKeys(state.listKeys, appendPath(intent.payload.name, index), updateListIndex)), {}, {
              // Update existing list keys
              [intent.payload.name]: selectedListKeys
            });
          }
        }
      }
      touchedFields = appendUniqueItem(touchedFields, intent.payload.name);
      return _objectSpread2(_objectSpread2({}, state), {}, {
        listKeys: listKeys,
        touchedFields
      });
    }
  },
  reorder: {
    validate(options) {
      return isPlainObject(options) && isString(options.name) && isNumber(options.from) && isNumber(options.to);
    },
    resolve(value, options) {
      var list = Array.from(getPathArray(value, options.name));
      reorderItems(list, options.from, options.to);
      return updatePathValue(value, options.name, list);
    },
    update(state, _ref5) {
      var {
        type,
        submission,
        intent
      } = _ref5;
      if (type === 'server') {
        return state;
      }
      var currentValue = submission.payload;
      var updateListIndex = createPathIndexUpdater(intent.payload.name, currentIndex => {
        if (intent.payload.from === intent.payload.to) {
          return currentIndex;
        }
        if (currentIndex === intent.payload.from) {
          return intent.payload.to;
        }
        if (intent.payload.from < intent.payload.to) {
          return currentIndex > intent.payload.from && currentIndex <= intent.payload.to ? currentIndex - 1 : currentIndex;
        }
        return currentIndex >= intent.payload.to && currentIndex < intent.payload.from ? currentIndex + 1 : currentIndex;
      });
      var touchedFields = appendUniqueItem(compactMap(state.touchedFields, updateListIndex), intent.payload.name);
      var keys = state.listKeys;

      // Update the keys only for client updates to avoid double updates if there is no client validation
      if (type === 'client') {
        var _state$listKeys$inten3;
        var listKeys = Array.from((_state$listKeys$inten3 = state.listKeys[intent.payload.name]) !== null && _state$listKeys$inten3 !== void 0 ? _state$listKeys$inten3 : getDefaultListKey(state.resetKey, currentValue, intent.payload.name));
        reorderItems(listKeys, intent.payload.from, intent.payload.to);
        keys = _objectSpread2(_objectSpread2({}, updateListKeys(state.listKeys, appendPath(intent.payload.name, intent.payload.from), updateListIndex)), {}, {
          // Update existing list keys
          [intent.payload.name]: listKeys
        });
      }
      return _objectSpread2(_objectSpread2({}, state), {}, {
        listKeys: keys,
        touchedFields
      });
    }
  }
};

export { applyIntent, deserializeIntent, insertItem, intentHandlers, removeItem, reorderItems, resolveIntent, serializeIntent, updateListKeys };
