// @conform-to/zod/dist/v4/future.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../_virtual/_rollupPluginBabelHelpers.js');
var coercion = require('./coercion.js');
var constraint = require('./constraint.js');
var format = require('./format.js');
var schema = require('./schema.js');

function defaultDate(text) {
  var date = new Date(shouldAppendUtcSuffix(text) ? "".concat(text, "Z") : text);
  if (isNaN(date.getTime())) {
    throw new Error('Invalid date');
  }
  return date;
}
function shouldAppendUtcSuffix(datetimeString) {
  if (datetimeString.includes(' ')) {
    return false;
  }
  var separatorIndex = datetimeString.indexOf('T');
  if (separatorIndex < 0) {
    return false;
  }
  var time = datetimeString.slice(separatorIndex + 1);
  return !(time.toUpperCase().endsWith('Z') || time.includes('+') || time.includes('-'));
}
function configureCoercion(config) {
  var _config$type$date, _config$type;
  return coercion.configureCoercion(_rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2({}, config), {}, {
    type: _rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2({}, config === null || config === void 0 ? void 0 : config.type), {}, {
      date: (_config$type$date = config === null || config === void 0 || (_config$type = config.type) === null || _config$type === void 0 ? void 0 : _config$type.date) !== null && _config$type$date !== void 0 ? _config$type$date : defaultDate
    })
  }));
}
var {
  coerceFormValue,
  coerceStructure
} = configureCoercion();

exports.getZodConstraint = constraint.getZodConstraint;
exports.formatResult = format.formatResult;
exports.getConstraints = schema.getConstraints;
exports.isSchema = schema.isSchema;
exports.coerceFormValue = coerceFormValue;
exports.coerceStructure = coerceStructure;
exports.configureCoercion = configureCoercion;
