// @conform-to/zod/dist/v3/future.mjs
import { objectSpread2 as _objectSpread2 } from '../_virtual/_rollupPluginBabelHelpers.mjs';
import { configureCoercion as configureCoercion$1 } from './coercion.mjs';
export { getZodConstraint } from './constraint.mjs';
export { formatResult } from './format.mjs';
export { getConstraints, isSchema } from './schema.mjs';

function defaultDate(text) {
  var date = new Date(shouldAppendUtcSuffix(text) ? "".concat(text, "Z") : text);
  if (isNaN(date.getTime())) {
    throw new Error('Invalid date');
  }
  return date;
}
function shouldAppendUtcSuffix(datetimeString) {
  if (datetimeString.includes(' ')) {
    return false;
  }
  var separatorIndex = datetimeString.indexOf('T');
  if (separatorIndex < 0) {
    return false;
  }
  var time = datetimeString.slice(separatorIndex + 1);
  return !(time.toUpperCase().endsWith('Z') || time.includes('+') || time.includes('-'));
}
function configureCoercion(config) {
  var _config$type$date, _config$type;
  return configureCoercion$1(_objectSpread2(_objectSpread2({}, config), {}, {
    type: _objectSpread2(_objectSpread2({}, config === null || config === void 0 ? void 0 : config.type), {}, {
      date: (_config$type$date = config === null || config === void 0 || (_config$type = config.type) === null || _config$type === void 0 ? void 0 : _config$type.date) !== null && _config$type$date !== void 0 ? _config$type$date : defaultDate
    })
  }));
}
var {
  coerceFormValue,
  coerceStructure
} = configureCoercion();

export { coerceFormValue, coerceStructure, configureCoercion };
