// vite-plugin-cjs-interop/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
let oxc_parser = require("oxc-parser");
let magic_string = require("magic-string");
magic_string = __toESM(magic_string);
let minimatch = require("minimatch");
//#region src/index.ts
const walker = import("estree-walker");
const CSS_LANGS_RE = /\.(css|less|sass|scss|styl|stylus|pcss|postcss|sss)(?:$|\?)/;
function cjsInterop(options) {
	const virtualModuleId = "virtual:cjs-dyn-import";
	const dependencies = Array.from(new Set(options.dependencies));
	const { client = false, apply = "both" } = options;
	let sourcemaps = false;
	const matchedDependencies = {};
	const matchesDependencies = (value) => {
		if (!(value in matchedDependencies)) matchedDependencies[value] = dependencies.some((dependency) => (0, minimatch.minimatch)(value, dependency));
		return matchedDependencies[value];
	};
	return {
		name: "cjs-interop",
		enforce: "post",
		apply: apply === "both" ? void 0 : apply,
		configResolved(config) {
			sourcemaps = !!config.build.sourcemap;
		},
		async transform(code, id, options) {
			if (!client && !options?.ssr) return;
			if (CSS_LANGS_RE.test(id)) return;
			const { program: ast } = await (0, oxc_parser.parse)(id, code);
			const toBeFixed = [];
			const preambles = [];
			let hasDynamicImportsToFix = false;
			const { walk } = await walker;
			walk(ast, { enter(node) {
				switch (node.type) {
					case "ImportDeclaration":
					case "ExportNamedDeclaration":
					case "ImportExpression":
						if (node.source && node.source.type === "Literal" && matchesDependencies(node.source.value)) {
							toBeFixed.push(node);
							if (node.type === "ImportExpression") hasDynamicImportsToFix = true;
						}
						break;
					default:
				}
			} });
			if (toBeFixed.length === 0) return;
			const bottomUpToBeFixed = toBeFixed.reverse();
			const ms = sourcemaps ? new magic_string.default(code) : null;
			let counter = 1;
			let specifierCounter = 1;
			for (const node of bottomUpToBeFixed) {
				let isNamespaceImport = false;
				if (node.type === "ImportExpression") {
					const insertion = ".then(__cjs_dyn_import__)";
					if (sourcemaps) ms.appendRight(node.end, insertion);
					else code = code.slice(0, node.end) + insertion + code.slice(node.end);
					continue;
				}
				if (node.type === "ExportNamedDeclaration") {
					const importDestructurings = [];
					const exportDestructurings = [];
					const name = `__cjsInterop${counter++}__`;
					let changed = false;
					let defaultExportSpecifier = null;
					for (const specifier of node.specifiers) if (specifier.type === "ExportSpecifier") {
						if (specifier.local.name === "default") {
							defaultExportSpecifier = specifier;
							continue;
						}
						changed = true;
						const specifierName = `__cjsInteropSpecifier${specifierCounter++}__`;
						importDestructurings.push(`${specifier.local.name}: ${specifierName}`);
						exportDestructurings.push(`${specifierName} as ${specifier.exported.name}`);
					} else throw new Error(`Unknown ExportNamedDeclaration type specifier: ${specifier.type}`);
					if (!changed) continue;
					if (defaultExportSpecifier) exportDestructurings.push(`${name} as ${defaultExportSpecifier.exported.name}}`);
					preambles.push(`const { ${importDestructurings.join(", ")} } = ${name}?.default?.__esModule ? ${name}.default : ${name};`);
					const replacement = [`import ${name} from ${JSON.stringify(node.source.value)};`, `export { ${exportDestructurings.join(", ")} };`].filter(Boolean).join("\n");
					if (ms) ms.overwrite(node.start, node.end, replacement);
					else code = code.slice(0, node.start) + replacement + code.slice(node.end);
					continue;
				}
				const destructurings = [];
				const name = `__cjsInterop${counter++}__`;
				let changed = false;
				for (const specifier of node.specifiers || []) if (specifier.type === "ImportDefaultSpecifier") {
					changed = true;
					destructurings.push(`default: ${specifier.local.name} = ${name}`);
				} else if (specifier.type === "ImportSpecifier") {
					changed = true;
					if (specifier.imported.name === specifier.local.name) destructurings.push(specifier.local.name);
					else destructurings.push(`${specifier.imported.name}: ${specifier.local.name}`);
				} else if (specifier.type === "ImportNamespaceSpecifier") {
					changed = true;
					isNamespaceImport = true;
					destructurings.push(specifier.local.name);
				}
				if (!changed) continue;
				if (!isNamespaceImport) preambles.push(`const { ${destructurings.join(", ")} } = ${name}?.default?.__esModule ? ${name}.default : ${name};`);
				else preambles.push(`const ${destructurings[0]} = ${name}?.default?.__esModule ? ${name}.default : ${name};`);
				const replacement = `import ${name} from ${JSON.stringify(node.source.value)};`;
				if (sourcemaps) ms.overwrite(node.start, node.end, replacement);
				else code = code.slice(0, node.start) + replacement + code.slice(node.end);
			}
			if (hasDynamicImportsToFix) {
				const importCompat = `import { __cjs_dyn_import__ } from "${virtualModuleId}";\n`;
				if (sourcemaps) ms.prepend(importCompat);
				else code = importCompat + code;
			}
			const preamble = preambles.reverse().join("\n") + "\n";
			if (sourcemaps) {
				ms.prepend(preamble);
				return {
					code: ms.toString(),
					map: ms.generateMap({ hires: true })
				};
			} else {
				code = preamble + code;
				return { code };
			}
		},
		resolveId(id) {
			if (id === virtualModuleId) return id;
		},
		load(id) {
			if (id === virtualModuleId) return `
// This file is generated by vite-plugin-cjs-interop
// The cache allows the same object to be returned for the same dynamic import
// however static imports are not affected and will therefore return a different
// object
const modCache = new WeakMap();

export function __cjs_dyn_import__(rawImport) {
  if (rawImport?.default?.__esModule) {
    return rawImport.default;
  }
  if (modCache.has(rawImport)) {
    return modCache.get(rawImport);
  }
  const source = rawImport?.default;
  if (source) {
    const mod = Object.create(rawImport);
    modCache.set(rawImport, mod);
    Object.keys(source)
      .filter(key => !Object.hasOwn(rawImport, key))
      .forEach(key => {
        Object.defineProperty(mod, key, { enumerable: true, get: () => source[key] })
      });
    return mod;
  }
  return rawImport;
}`;
		}
	};
}
//#endregion
exports.cjsInterop = cjsInterop;
