// @conform-to/dom/dist/future/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var formdata = require('../formdata.js');
var util = require('../util.js');
var dom = require('../dom.js');
var standardSchema = require('../standard-schema.js');



exports.DEFAULT_INTENT_NAME = formdata.DEFAULT_INTENT_NAME;
exports.appendPath = formdata.appendPath;
exports.formatPath = formdata.formatPath;
exports.getFieldValue = formdata.getFieldValue;
exports.getFormData = formdata.getFormData;
exports.getPathValue = formdata.getPathValue;
exports.getRelativePath = formdata.getRelativePath;
exports.isDirty = formdata.isDirty;
exports.normalize = formdata.normalize;
exports.parsePath = formdata.parsePath;
exports.parseSubmission = formdata.parseSubmission;
exports.report = formdata.report;
exports.serialize = formdata.serialize;
exports.setPathValue = formdata.setPathValue;
exports.deepEqual = util.deepEqual;
exports.isPlainObject = util.isPlainObject;
exports.blur = dom.blur;
exports.change = dom.change;
exports.createFileList = dom.createFileList;
exports.createGlobalFormsObserver = dom.createGlobalFormsObserver;
exports.dispatchInternalUpdateEvent = dom.dispatchInternalUpdateEvent;
exports.focus = dom.focus;
exports.getFormAction = dom.getFormAction;
exports.getFormEncType = dom.getFormEncType;
exports.getFormMethod = dom.getFormMethod;
exports.isFieldElement = dom.isFieldElement;
exports.isGlobalInstance = dom.isGlobalInstance;
exports.requestIntent = dom.requestIntent;
exports.requestSubmit = dom.requestSubmit;
exports.updateField = dom.updateField;
exports.formatIssues = standardSchema.formatIssues;
