// @reach/checkbox/dist/reach-checkbox.cjs.dev.js
"use strict";
/**
  * @reach/checkbox v0.18.0
  *
  * Copyright (c) 2018-2022, React Training LLC
  *
  * This source code is licensed under the MIT license found in the
  * LICENSE.md file in the root directory of this source tree.
  *
  * @license MIT
  */

var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/reach-checkbox.ts
var reach_checkbox_exports = {};
__export(reach_checkbox_exports, {
  CustomCheckbox: () => CustomCheckbox,
  CustomCheckboxContainer: () => CustomCheckboxContainer,
  CustomCheckboxInput: () => CustomCheckboxInput,
  MixedCheckbox: () => MixedCheckbox,
  MixedCheckboxEvents: () => MixedCheckboxEvents,
  MixedCheckboxStates: () => MixedCheckboxStates,
  internal_checkedPropToStateValue: () => checkedPropToStateValue,
  useMixedCheckbox: () => useMixedCheckbox
});
module.exports = __toCommonJS(reach_checkbox_exports);

// src/custom.tsx
var React2 = __toESM(require("react"));
var import_utils2 = require("@reach/utils");

// src/mixed.tsx
var React = __toESM(require("react"));
var import_utils = require("@reach/utils");
var import_machine = require("@reach/machine");
var DEBUG = false;
var MixedCheckboxStates = /* @__PURE__ */ ((MixedCheckboxStates2) => {
  MixedCheckboxStates2["Checked"] = "checked";
  MixedCheckboxStates2["Mixed"] = "mixed";
  MixedCheckboxStates2["Unchecked"] = "unchecked";
  return MixedCheckboxStates2;
})(MixedCheckboxStates || {});
var MixedCheckboxEvents = /* @__PURE__ */ ((MixedCheckboxEvents2) => {
  MixedCheckboxEvents2["GetDerivedData"] = "GET_DERIVED_DATA";
  MixedCheckboxEvents2["Mount"] = "MOUNT";
  MixedCheckboxEvents2["Set"] = "SET";
  MixedCheckboxEvents2["Reset"] = "RESET";
  MixedCheckboxEvents2["Toggle"] = "TOGGLE";
  MixedCheckboxEvents2["Unmount"] = "UNMOUNT";
  return MixedCheckboxEvents2;
})(MixedCheckboxEvents || {});
function checkToggleAllowed(data) {
  return !!(data && !data.isControlled && !data.disabled);
}
function getCheckSetCondition(state) {
  return function(data, event) {
    return data && data.isControlled && event.state === state;
  };
}
function getCheckResetCondition(state) {
  return function(data, event) {
    return data && !data.isControlled && data.initialState === state;
  };
}
var assignRefs = (0, import_machine.assign)((data, event) => {
  return __spreadProps(__spreadValues({}, data), {
    refs: event.refs
  });
});
var commonEvents = {
  ["MOUNT" /* Mount */]: {
    actions: assignRefs
  },
  ["GET_DERIVED_DATA" /* GetDerivedData */]: {
    actions: [
      assignRefs,
      (0, import_machine.assign)((data, event) => {
        return __spreadValues(__spreadValues({}, data), event.data);
      })
    ]
  },
  ["SET" /* Set */]: [
    {
      target: "checked" /* Checked */,
      cond: getCheckSetCondition("checked" /* Checked */)
    },
    {
      target: "unchecked" /* Unchecked */,
      cond: getCheckSetCondition("unchecked" /* Unchecked */)
    },
    {
      target: "mixed" /* Mixed */,
      cond: getCheckSetCondition("mixed" /* Mixed */)
    }
  ],
  ["RESET" /* Reset */]: [
    {
      target: "checked" /* Checked */,
      cond: getCheckResetCondition("checked" /* Checked */)
    },
    {
      target: "unchecked" /* Unchecked */,
      cond: getCheckResetCondition("unchecked" /* Unchecked */)
    },
    {
      target: "mixed" /* Mixed */,
      cond: getCheckResetCondition("mixed" /* Mixed */)
    }
  ]
};
var createMachineDefinition = (initialState, props) => ({
  id: "mixed-checkbox",
  initial: initialState,
  context: {
    disabled: props.disabled,
    initialState,
    isControlled: props.isControlled,
    refs: {
      input: null
    }
  },
  states: {
    ["unchecked" /* Unchecked */]: {
      entry: assignRefs,
      on: __spreadValues({
        ["TOGGLE" /* Toggle */]: {
          target: "checked" /* Checked */,
          cond: checkToggleAllowed
        }
      }, commonEvents)
    },
    ["checked" /* Checked */]: {
      entry: assignRefs,
      on: __spreadValues({
        ["TOGGLE" /* Toggle */]: {
          target: "unchecked" /* Unchecked */,
          cond: checkToggleAllowed
        }
      }, commonEvents)
    },
    ["mixed" /* Mixed */]: {
      entry: assignRefs,
      on: __spreadValues({
        ["TOGGLE" /* Toggle */]: {
          target: "checked" /* Checked */,
          cond: checkToggleAllowed
        }
      }, commonEvents)
    }
  }
});
var MixedCheckbox = React.forwardRef(function MixedCheckbox2(_a, forwardedRef) {
  var _b = _a, { as: Comp = "input", checked, defaultChecked, disabled, onChange } = _b, props = __objRest(_b, ["as", "checked", "defaultChecked", "disabled", "onChange"]);
  let ownRef = React.useRef(null);
  let ref = (0, import_utils.useComposedRefs)(forwardedRef, ownRef);
  let [inputProps] = useMixedCheckbox(ownRef, {
    onChange,
    checked,
    defaultChecked,
    disabled
  }, "MixedCheckbox");
  return /* @__PURE__ */ React.createElement(Comp, __spreadProps(__spreadValues(__spreadValues({}, props), inputProps), {
    "data-reach-mixed-checkbox": "",
    ref
  }));
});
MixedCheckbox.displayName = "MixedCheckbox";
function useMixedCheckbox(inputRef, args, deprecated_callerName) {
  let callerName = (args == null ? void 0 : args.callerName) || deprecated_callerName || "useMixedCheckbox";
  let {
    checked: controlledChecked,
    defaultChecked,
    disabled,
    onChange,
    onClick
  } = args || {};
  let isControlled = controlledChecked != null;
  let machine = (0, import_machine.useCreateMachine)(createMachineDefinition(checkedPropToStateValue(isControlled ? controlledChecked : defaultChecked), {
    disabled: !!disabled,
    isControlled
  }));
  let [current, send] = (0, import_machine.useMachine)(machine, { input: inputRef }, DEBUG);
  let props = {
    checked: stateValueToChecked(current.value),
    disabled: !!disabled,
    onChange: (0, import_utils.composeEventHandlers)(onChange, handleChange),
    onClick: (0, import_utils.composeEventHandlers)(onClick, handleClick),
    type: "checkbox"
  };
  if (current.value === "mixed" /* Mixed */) {
    props["aria-checked"] = "mixed";
  }
  let contextData = {
    checked: current.value === "checked" /* Checked */ ? true : current.value === "mixed" /* Mixed */ ? "mixed" : false
  };
  function handleChange() {
    if (!isControlled) {
      send("TOGGLE" /* Toggle */);
    }
  }
  function handleClick() {
    syncDomNodeWithState();
  }
  function syncDomNodeWithState() {
    if (inputRef.current) {
      inputRef.current.indeterminate = current.value === "mixed" /* Mixed */;
    }
  }
  if (true) {
    let { current: wasControlled } = React.useRef(isControlled);
    React.useEffect(() => {
      if (!inputRef.current) {
        console.warn(`A ref was not assigned to an input element in ${callerName}.`);
      }
    }, [inputRef, callerName]);
    React.useEffect(() => {
      if (!isControlled && wasControlled) {
        console.warn(`${callerName} is changing from controlled to uncontrolled. ${callerName} should not switch from controlled to uncontrolled (or vice versa). Decide between using a controlled or uncontrolled ${callerName} for the lifetime of the component. Check the \`checked\` prop being passed in.`);
      }
      if (isControlled && !wasControlled) {
        console.warn(`${callerName} is changing from uncontrolled to controlled. ${callerName} should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled ${callerName} for the lifetime of the component. Check the \`checked\` prop being passed in.`);
      }
    }, [callerName, isControlled, wasControlled]);
  }
  React.useEffect(() => {
    if (isControlled) {
      send({
        type: "SET" /* Set */,
        state: checkedPropToStateValue(controlledChecked)
      });
    }
  }, [isControlled, controlledChecked, send]);
  React.useEffect(() => {
    var _a;
    let form = (_a = inputRef.current) == null ? void 0 : _a.form;
    if (!form)
      return;
    function handleReset(event) {
      send({ type: "RESET" /* Reset */ });
    }
    form.addEventListener("reset", handleReset);
    return () => {
      form == null ? void 0 : form.removeEventListener("reset", handleReset);
    };
  }, [inputRef, isControlled, send]);
  (0, import_utils.useIsomorphicLayoutEffect)(syncDomNodeWithState);
  React.useEffect(() => {
    send({
      type: "GET_DERIVED_DATA" /* GetDerivedData */,
      data: {
        disabled,
        isControlled
      }
    });
  }, [disabled, isControlled, send]);
  return [props, contextData];
}
function checkedPropToStateValue(checked) {
  return checked === true ? "checked" /* Checked */ : checked === "mixed" ? "mixed" /* Mixed */ : "unchecked" /* Unchecked */;
}
function stateValueToChecked(state) {
  return state === "checked" /* Checked */ ? true : false;
}

// src/custom.tsx
var [CustomCheckboxContextProvider, useCustomCheckboxContext] = (0, import_utils2.createContext)("CustomCheckboxContext", {});
var CustomCheckboxContainer = React2.forwardRef(function CustomCheckboxContainer2(_a, forwardedRef) {
  var _b = _a, {
    as: Comp = "span",
    checked: controlledChecked,
    children,
    defaultChecked,
    disabled,
    onClick,
    onChange,
    __componentName = "CustomCheckboxContainer"
  } = _b, props = __objRest(_b, [
    "as",
    "checked",
    "children",
    "defaultChecked",
    "disabled",
    "onClick",
    "onChange",
    "__componentName"
  ]);
  let inputRef = React2.useRef(null);
  let [inputProps, stateData] = useMixedCheckbox(inputRef, {
    defaultChecked,
    checked: controlledChecked,
    disabled,
    onChange
  }, __componentName);
  let [focused, setFocused] = React2.useState(false);
  function handleClick() {
    window.requestAnimationFrame(() => {
      inputRef.current && inputRef.current.focus();
    });
  }
  let context = {
    defaultChecked,
    disabled,
    focused,
    inputProps,
    inputRef,
    setFocused
  };
  return /* @__PURE__ */ React2.createElement(CustomCheckboxContextProvider, __spreadValues({}, context), /* @__PURE__ */ React2.createElement(Comp, __spreadProps(__spreadValues({}, props), {
    ref: forwardedRef,
    "data-reach-custom-checkbox-container": "",
    "data-focused": focused ? "" : void 0,
    "data-state": checkedPropToStateValue(stateData.checked),
    onClick: (0, import_utils2.composeEventHandlers)(onClick, handleClick)
  }), (0, import_utils2.isFunction)(children) ? children({
    checked: stateData.checked,
    inputRef,
    focused
  }) : children));
});
CustomCheckboxContainer.displayName = "CustomCheckboxContainer";
var CustomCheckboxInput = React2.forwardRef(function CustomCheckboxInput2(_a, forwardedRef) {
  var _b = _a, { as: Comp = "input", onBlur, onFocus } = _b, props = __objRest(_b, ["as", "onBlur", "onFocus"]);
  let { focused, inputProps, inputRef, setFocused } = useCustomCheckboxContext("CustomCheckboxInput");
  let ref = (0, import_utils2.useComposedRefs)(forwardedRef, inputRef);
  let mounted = React2.useRef(true);
  function handleBlur() {
    window.requestAnimationFrame(() => {
      if (mounted.current) {
        setFocused(false);
      }
    });
  }
  function handleFocus() {
    window.requestAnimationFrame(() => {
      if (mounted.current) {
        setFocused(true);
      }
    });
  }
  React2.useEffect(() => () => void (mounted.current = false), []);
  return /* @__PURE__ */ React2.createElement(Comp, __spreadProps(__spreadValues(__spreadValues({}, props), inputProps), {
    ref,
    type: "checkbox",
    "data-reach-custom-checkbox-input": "",
    "data-focused": focused ? "" : void 0,
    onBlur: (0, import_utils2.composeEventHandlers)(onBlur, handleBlur),
    onFocus: (0, import_utils2.composeEventHandlers)(onFocus, handleFocus)
  }));
});
CustomCheckboxInput.displayName = "CustomCheckboxInput";
var CustomCheckbox = React2.forwardRef(function CustomCheckbox2(_a, forwardedRef) {
  var _b = _a, { children, id, name, value } = _b, props = __objRest(_b, ["children", "id", "name", "value"]);
  return /* @__PURE__ */ React2.createElement(CustomCheckboxContainer, __spreadProps(__spreadValues({}, props), {
    "data-reach-custom-checkbox": "",
    __componentName: "CustomCheckbox"
  }), /* @__PURE__ */ React2.createElement(CustomCheckboxInput, {
    id,
    name,
    ref: forwardedRef,
    value
  }), children);
});
CustomCheckbox.displayName = "CustomCheckbox";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CustomCheckbox,
  CustomCheckboxContainer,
  CustomCheckboxInput,
  MixedCheckbox,
  MixedCheckboxEvents,
  MixedCheckboxStates,
  internal_checkedPropToStateValue,
  useMixedCheckbox
});
