// @reach/tabs/dist/reach-tabs.cjs.dev.js
"use strict";
/**
  * @reach/tabs v0.18.0
  *
  * Copyright (c) 2018-2022, React Training LLC
  *
  * This source code is licensed under the MIT license found in the
  * LICENSE.md file in the root directory of this source tree.
  *
  * @license MIT
  */

var __create = Object.create;
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/reach-tabs.tsx
var reach_tabs_exports = {};
__export(reach_tabs_exports, {
  Tab: () => Tab,
  TabList: () => TabList,
  TabPanel: () => TabPanel,
  TabPanels: () => TabPanels,
  Tabs: () => Tabs,
  TabsKeyboardActivation: () => TabsKeyboardActivation,
  TabsOrientation: () => TabsOrientation,
  useTabsContext: () => useTabsContext
});
module.exports = __toCommonJS(reach_tabs_exports);
var React = __toESM(require("react"));
var import_descendants = require("@reach/descendants");
var import_utils = require("@reach/utils");
var import_auto_id = require("@reach/auto-id");
var TabsDescendantsContext = (0, import_descendants.createDescendantContext)("TabsDescendantsContext");
var TabPanelDescendantsContext = (0, import_descendants.createDescendantContext)("TabPanelDescendantsContext");
var [TabsProvider, useTabsCtx] = (0, import_utils.createContext)("Tabs");
var TabsKeyboardActivation = /* @__PURE__ */ ((TabsKeyboardActivation2) => {
  TabsKeyboardActivation2["Auto"] = "auto";
  TabsKeyboardActivation2["Manual"] = "manual";
  return TabsKeyboardActivation2;
})(TabsKeyboardActivation || {});
var TabsOrientation = /* @__PURE__ */ ((TabsOrientation2) => {
  TabsOrientation2["Horizontal"] = "horizontal";
  TabsOrientation2["Vertical"] = "vertical";
  return TabsOrientation2;
})(TabsOrientation || {});
var Tabs = React.forwardRef((_a, ref) => {
  var _b = _a, {
    as: Comp = "div",
    children,
    defaultIndex,
    orientation = "horizontal" /* Horizontal */,
    index: controlledIndex = void 0,
    keyboardActivation = "auto" /* Auto */,
    onChange,
    readOnly = false
  } = _b, props = __objRest(_b, [
    "as",
    "children",
    "defaultIndex",
    "orientation",
    "index",
    "keyboardActivation",
    "onChange",
    "readOnly"
  ]);
  let { current: isControlled } = React.useRef(controlledIndex !== void 0);
  let id = (0, import_auto_id.useId)(props.id || "tabs");
  let selectedPanelRef = React.useRef(null);
  let isRTL = React.useRef(false);
  let [selectedIndex, setSelectedIndex] = (0, import_utils.useControlledState)({
    controlledValue: controlledIndex,
    defaultValue: defaultIndex != null ? defaultIndex : 0,
    calledFrom: "Tabs"
  });
  let [focusedIndex, setFocusedIndex] = React.useState(-1);
  let [tabs, setTabs] = (0, import_descendants.useDescendantsInit)();
  let onFocusPanel = React.useCallback(() => {
    if (selectedPanelRef.current && (0, import_utils.isFunction)(selectedPanelRef.current.focus)) {
      selectedPanelRef.current.focus();
    }
  }, []);
  let onSelectTab = React.useCallback((index) => {
    onChange && onChange(index);
    setSelectedIndex(index);
  }, [onChange, setSelectedIndex]);
  let onSelectTabWithKeyboard = React.useCallback((index) => {
    var _a2;
    let tabElement = (_a2 = tabs[index]) == null ? void 0 : _a2.element;
    let doc = (0, import_utils.getOwnerDocument)(tabElement);
    if (keyboardActivation === "auto" /* Auto */) {
      onChange && onChange(index);
      setSelectedIndex(index);
    }
    if (tabElement && tabElement !== doc.activeElement && (0, import_utils.isFunction)(tabElement.focus)) {
      tabElement.focus();
    }
  }, [keyboardActivation, onChange, setSelectedIndex, tabs]);
  return /* @__PURE__ */ React.createElement(import_descendants.DescendantProvider, {
    context: TabsDescendantsContext,
    items: tabs,
    set: setTabs
  }, /* @__PURE__ */ React.createElement(TabsProvider, {
    focusedIndex,
    id,
    isControlled,
    isRTL,
    keyboardActivation,
    onFocusPanel,
    onSelectTab: readOnly ? import_utils.noop : onSelectTab,
    onSelectTabWithKeyboard: readOnly ? import_utils.noop : onSelectTabWithKeyboard,
    orientation,
    selectedIndex,
    selectedPanelRef,
    setFocusedIndex,
    setSelectedIndex
  }, /* @__PURE__ */ React.createElement(Comp, __spreadProps(__spreadValues({}, props), {
    ref,
    "data-reach-tabs": "",
    "data-orientation": orientation,
    id: props.id
  }), (0, import_utils.isFunction)(children) ? children({ focusedIndex, id, selectedIndex }) : children)));
});
Tabs.displayName = "Tabs";
var TabListImpl = React.forwardRef((_a, forwardedRef) => {
  var _b = _a, { children, as: Comp = "div", onKeyDown } = _b, props = __objRest(_b, ["children", "as", "onKeyDown"]);
  let {
    focusedIndex,
    isControlled,
    isRTL,
    keyboardActivation,
    onSelectTabWithKeyboard,
    orientation,
    selectedIndex,
    setSelectedIndex
  } = useTabsCtx("TabList");
  let tabs = (0, import_descendants.useDescendants)(TabsDescendantsContext);
  let ownRef = React.useRef(null);
  let ref = (0, import_utils.useComposedRefs)(forwardedRef, ownRef);
  React.useEffect(() => {
    if (ownRef.current && (ownRef.current.ownerDocument && ownRef.current.ownerDocument.dir === "rtl" || (0, import_utils.getComputedStyle)(ownRef.current, "direction") === "rtl")) {
      isRTL.current = true;
    }
  }, [isRTL]);
  let handleKeyDown = (0, import_utils.composeEventHandlers)(onKeyDown, (0, import_descendants.useDescendantKeyDown)(TabsDescendantsContext, {
    currentIndex: keyboardActivation === "manual" /* Manual */ ? focusedIndex : selectedIndex,
    orientation,
    rotate: true,
    callback: onSelectTabWithKeyboard,
    filter: (tab) => !tab.disabled,
    rtl: isRTL.current
  }));
  (0, import_utils.useIsomorphicLayoutEffect)(() => {
    var _a2;
    if (!isControlled && boolOrBoolString((_a2 = tabs[selectedIndex]) == null ? void 0 : _a2.disabled)) {
      let next = tabs.find((tab) => !tab.disabled);
      if (next) {
        setSelectedIndex(next.index);
      }
    }
  }, [tabs, isControlled, selectedIndex, setSelectedIndex]);
  return /* @__PURE__ */ React.createElement(Comp, __spreadProps(__spreadValues({
    role: "tablist",
    "aria-orientation": orientation
  }, props), {
    "data-reach-tab-list": "",
    ref,
    onKeyDown: handleKeyDown
  }), children);
});
TabListImpl.displayName = "TabList";
var TabList = React.memo(TabListImpl);
TabList.displayName = "TabList";
var Tab = React.forwardRef((_a, forwardedRef) => {
  var _b = _a, {
    isSelected: _,
    children,
    as: Comp = "button",
    index: indexProp,
    disabled,
    onBlur,
    onFocus
  } = _b, props = __objRest(_b, [
    "isSelected",
    "children",
    "as",
    "index",
    "disabled",
    "onBlur",
    "onFocus"
  ]);
  let {
    id: tabsId,
    onSelectTab,
    orientation,
    selectedIndex,
    setFocusedIndex
  } = useTabsCtx("Tab");
  let ownRef = React.useRef(null);
  let [element, handleRefSet] = (0, import_utils.useStatefulRefValue)(ownRef, null);
  let ref = (0, import_utils.useComposedRefs)(forwardedRef, handleRefSet);
  let descendant = React.useMemo(() => {
    return {
      element,
      disabled: !!disabled
    };
  }, [disabled, element]);
  let index = (0, import_descendants.useDescendant)(descendant, TabsDescendantsContext, indexProp);
  let htmlType = Comp === "button" && props.type == null ? "button" : props.type;
  let isSelected = index === selectedIndex;
  function onSelect() {
    onSelectTab(index);
  }
  return /* @__PURE__ */ React.createElement(Comp, __spreadProps(__spreadValues({
    "aria-controls": (0, import_utils.makeId)(tabsId, "panel", index),
    "aria-disabled": disabled,
    "aria-selected": isSelected,
    role: "tab",
    tabIndex: isSelected ? 0 : -1
  }, props), {
    ref,
    "data-reach-tab": "",
    "data-orientation": orientation,
    "data-selected": isSelected ? "" : void 0,
    disabled,
    id: (0, import_utils.makeId)(tabsId, "tab", index),
    onClick: onSelect,
    onFocus: (0, import_utils.composeEventHandlers)(onFocus, () => {
      setFocusedIndex(index);
    }),
    onBlur: (0, import_utils.composeEventHandlers)(onBlur, () => {
      setFocusedIndex(-1);
    }),
    type: htmlType
  }), children);
});
Tab.displayName = "Tab";
var TabPanelsImpl = React.forwardRef((_a, forwardedRef) => {
  var _b = _a, { children, as: Comp = "div" } = _b, props = __objRest(_b, ["children", "as"]);
  let ownRef = React.useRef();
  let ref = (0, import_utils.useComposedRefs)(ownRef, forwardedRef);
  let [tabPanels, setTabPanels] = (0, import_descendants.useDescendantsInit)();
  return /* @__PURE__ */ React.createElement(import_descendants.DescendantProvider, {
    context: TabPanelDescendantsContext,
    items: tabPanels,
    set: setTabPanels
  }, /* @__PURE__ */ React.createElement(Comp, __spreadProps(__spreadValues({}, props), {
    ref,
    "data-reach-tab-panels": ""
  }), children));
});
TabPanelsImpl.displayName = "TabPanels";
var TabPanels = React.memo(TabPanelsImpl);
TabPanels.displayName = "TabPanels";
var TabPanel = React.forwardRef((_a, forwardedRef) => {
  var _b = _a, {
    children,
    "aria-label": ariaLabel,
    as: Comp = "div",
    index: indexProp
  } = _b, props = __objRest(_b, [
    "children",
    "aria-label",
    "as",
    "index"
  ]);
  let {
    selectedPanelRef,
    selectedIndex,
    id: tabsId
  } = useTabsCtx("TabPanel");
  let ownRef = React.useRef(null);
  let [element, handleRefSet] = (0, import_utils.useStatefulRefValue)(ownRef, null);
  let descendant = React.useMemo(() => ({ element }), [element]);
  let index = (0, import_descendants.useDescendant)(descendant, TabPanelDescendantsContext, indexProp);
  let id = (0, import_utils.makeId)(tabsId, "panel", index);
  let isSelected = index === selectedIndex;
  let readyToHide = React.useRef(false);
  let hidden = readyToHide.current ? !isSelected : false;
  React.useEffect(() => {
    readyToHide.current = true;
  }, []);
  let ref = (0, import_utils.useComposedRefs)(forwardedRef, handleRefSet, isSelected ? selectedPanelRef : null);
  return /* @__PURE__ */ React.createElement(Comp, __spreadProps(__spreadValues({
    "aria-labelledby": (0, import_utils.makeId)(tabsId, "tab", index),
    hidden,
    role: "tabpanel",
    tabIndex: isSelected ? 0 : -1
  }, props), {
    ref,
    "data-reach-tab-panel": "",
    id
  }), children);
});
TabPanel.displayName = "TabPanel";
function useTabsContext() {
  let { focusedIndex, id, selectedIndex } = useTabsCtx("useTabsContext");
  return React.useMemo(() => ({
    focusedIndex,
    id,
    selectedIndex
  }), [focusedIndex, id, selectedIndex]);
}
function boolOrBoolString(value) {
  return value === "true" ? true : (0, import_utils.isBoolean)(value) ? value : false;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Tab,
  TabList,
  TabPanel,
  TabPanels,
  Tabs,
  TabsKeyboardActivation,
  TabsOrientation,
  useTabsContext
});
