// @tanstack/react-hotkeys/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_HotkeysProvider = require('./HotkeysProvider.cjs');
const require_useHotkey = require('./useHotkey.cjs');
const require_useHeldKeys = require('./useHeldKeys.cjs');
const require_useHeldKeyCodes = require('./useHeldKeyCodes.cjs');
const require_useKeyHold = require('./useKeyHold.cjs');
const require_useHotkeySequence = require('./useHotkeySequence.cjs');
const require_useHotkeyRecorder = require('./useHotkeyRecorder.cjs');

exports.HotkeysProvider = require_HotkeysProvider.HotkeysProvider;
exports.useDefaultHotkeysOptions = require_HotkeysProvider.useDefaultHotkeysOptions;
exports.useHeldKeyCodes = require_useHeldKeyCodes.useHeldKeyCodes;
exports.useHeldKeys = require_useHeldKeys.useHeldKeys;
exports.useHotkey = require_useHotkey.useHotkey;
exports.useHotkeyRecorder = require_useHotkeyRecorder.useHotkeyRecorder;
exports.useHotkeySequence = require_useHotkeySequence.useHotkeySequence;
exports.useHotkeysContext = require_HotkeysProvider.useHotkeysContext;
exports.useKeyHold = require_useKeyHold.useKeyHold;
var _tanstack_hotkeys = require("@tanstack/hotkeys");
Object.keys(_tanstack_hotkeys).forEach(function (k) {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return _tanstack_hotkeys[k]; }
  });
});
