// @tanstack/react-hotkeys/dist/useHotkey.cjs
const require_runtime = require('./_virtual/_rolldown/runtime.cjs');
const require_HotkeysProvider = require('./HotkeysProvider.cjs');
let _tanstack_hotkeys = require("@tanstack/hotkeys");
let react = require("react");

//#region src/useHotkey.ts
/**
* React hook for registering a keyboard hotkey.
*
* Uses the singleton HotkeyManager for efficient event handling.
* The callback receives both the keyboard event and a context object
* containing the hotkey string and parsed hotkey.
*
* This hook syncs the callback and options on every render to avoid
* stale closures. This means
* callbacks that reference React state will always have access to
* the latest values.
*
* @param hotkey - The hotkey string (e.g., 'Mod+S', 'Escape') or RawHotkey object (supports `mod` for cross-platform)
* @param callback - The function to call when the hotkey is pressed
* @param options - Options for the hotkey behavior
*
* @example
* ```tsx
* function SaveButton() {
*   const [count, setCount] = useState(0)
*
*   // Callback always has access to latest count value
*   useHotkey('Mod+S', (event, { hotkey }) => {
*     console.log(`Save triggered, count is ${count}`)
*     handleSave()
*   })
*
*   return <button onClick={() => setCount(c => c + 1)}>Count: {count}</button>
* }
* ```
*
* @example
* ```tsx
* function Modal({ isOpen, onClose }) {
*   // enabled option is synced on every render
*   useHotkey('Escape', () => {
*     onClose()
*   }, { enabled: isOpen })
*
*   if (!isOpen) return null
*   return <div className="modal">...</div>
* }
* ```
*
* @example
* ```tsx
* function Editor() {
*   const editorRef = useRef<HTMLDivElement>(null)
*
*   // Scoped to a specific element
*   useHotkey('Mod+S', () => {
*     save()
*   }, { target: editorRef })
*
*   return <div ref={editorRef}>...</div>
* }
* ```
*/
function useHotkey(hotkey, callback, options = {}) {
	const mergedOptions = {
		...require_HotkeysProvider.useDefaultHotkeysOptions().hotkey,
		...options
	};
	const manager = (0, _tanstack_hotkeys.getHotkeyManager)();
	const registrationRef = (0, react.useRef)(null);
	const callbackRef = (0, react.useRef)(callback);
	const optionsRef = (0, react.useRef)(mergedOptions);
	const managerRef = (0, react.useRef)(manager);
	callbackRef.current = callback;
	optionsRef.current = mergedOptions;
	managerRef.current = manager;
	const prevTargetRef = (0, react.useRef)(null);
	const prevHotkeyRef = (0, react.useRef)(null);
	const platform = mergedOptions.platform ?? (0, _tanstack_hotkeys.detectPlatform)();
	const hotkeyString = typeof hotkey === "string" ? hotkey : (0, _tanstack_hotkeys.formatHotkey)((0, _tanstack_hotkeys.rawHotkeyToParsedHotkey)(hotkey, platform));
	const { target: _target, ...optionsWithoutTarget } = mergedOptions;
	(0, react.useEffect)(() => {
		const resolvedTarget = isRef(optionsRef.current.target) ? optionsRef.current.target.current : optionsRef.current.target ?? (typeof document !== "undefined" ? document : null);
		if (!resolvedTarget) return;
		const targetChanged = prevTargetRef.current !== null && prevTargetRef.current !== resolvedTarget;
		const hotkeyChanged = prevHotkeyRef.current !== null && prevHotkeyRef.current !== hotkeyString;
		if (registrationRef.current?.isActive && (targetChanged || hotkeyChanged)) {
			registrationRef.current.unregister();
			registrationRef.current = null;
		}
		if (!registrationRef.current || !registrationRef.current.isActive) registrationRef.current = managerRef.current.register(hotkeyString, callbackRef.current, {
			...optionsRef.current,
			target: resolvedTarget
		});
		prevTargetRef.current = resolvedTarget;
		prevHotkeyRef.current = hotkeyString;
		return () => {
			if (registrationRef.current?.isActive) {
				registrationRef.current.unregister();
				registrationRef.current = null;
			}
		};
	}, [hotkeyString, options.enabled]);
	if (registrationRef.current?.isActive) {
		registrationRef.current.callback = callback;
		registrationRef.current.setOptions(optionsWithoutTarget);
	}
}
/**
* Type guard to check if a value is a React ref-like object.
*/
function isRef(value) {
	return value !== null && typeof value === "object" && "current" in value;
}

//#endregion
exports.useHotkey = useHotkey;
