// @tanstack/hotkeys/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_constants = require('./constants.cjs');
const require_parse = require('./parse.cjs');
const require_format = require('./format.cjs');
const require_match = require('./match.cjs');
const require_hotkey_manager = require('./hotkey-manager.cjs');
const require_key_state_tracker = require('./key-state-tracker.cjs');
const require_recorder = require('./recorder.cjs');
const require_sequence_manager = require('./sequence-manager.cjs');
const require_validate = require('./validate.cjs');

exports.ALL_KEYS = require_constants.ALL_KEYS;
exports.EDITING_KEYS = require_constants.EDITING_KEYS;
exports.FUNCTION_KEYS = require_constants.FUNCTION_KEYS;
exports.HotkeyManager = require_hotkey_manager.HotkeyManager;
exports.HotkeyRecorder = require_recorder.HotkeyRecorder;
exports.KEY_DISPLAY_SYMBOLS = require_constants.KEY_DISPLAY_SYMBOLS;
exports.KeyStateTracker = require_key_state_tracker.KeyStateTracker;
exports.LETTER_KEYS = require_constants.LETTER_KEYS;
exports.MAC_MODIFIER_SYMBOLS = require_constants.MAC_MODIFIER_SYMBOLS;
exports.MODIFIER_ALIASES = require_constants.MODIFIER_ALIASES;
exports.MODIFIER_KEYS = require_constants.MODIFIER_KEYS;
exports.MODIFIER_ORDER = require_constants.MODIFIER_ORDER;
exports.NAVIGATION_KEYS = require_constants.NAVIGATION_KEYS;
exports.NUMBER_KEYS = require_constants.NUMBER_KEYS;
exports.PUNCTUATION_CODE_MAP = require_constants.PUNCTUATION_CODE_MAP;
exports.PUNCTUATION_KEYS = require_constants.PUNCTUATION_KEYS;
exports.STANDARD_MODIFIER_LABELS = require_constants.STANDARD_MODIFIER_LABELS;
exports.SequenceManager = require_sequence_manager.SequenceManager;
exports.assertValidHotkey = require_validate.assertValidHotkey;
exports.checkHotkey = require_validate.checkHotkey;
exports.convertToModFormat = require_parse.convertToModFormat;
exports.createHotkeyHandler = require_match.createHotkeyHandler;
exports.createMultiHotkeyHandler = require_match.createMultiHotkeyHandler;
exports.createSequenceMatcher = require_sequence_manager.createSequenceMatcher;
exports.detectPlatform = require_constants.detectPlatform;
exports.formatForDisplay = require_format.formatForDisplay;
exports.formatHotkey = require_format.formatHotkey;
exports.formatHotkeySequence = require_format.formatHotkeySequence;
exports.formatKeyForDebuggingDisplay = require_format.formatKeyForDebuggingDisplay;
exports.formatWithLabels = require_format.formatWithLabels;
exports.getHotkeyManager = require_hotkey_manager.getHotkeyManager;
exports.getKeyStateTracker = require_key_state_tracker.getKeyStateTracker;
exports.getSequenceManager = require_sequence_manager.getSequenceManager;
exports.hasNonModifierKey = require_parse.hasNonModifierKey;
exports.isModifier = require_parse.isModifier;
exports.isModifierKey = require_parse.isModifierKey;
exports.isSingleLetterKey = require_constants.isSingleLetterKey;
exports.keyboardEventToHotkey = require_parse.keyboardEventToHotkey;
exports.matchesKeyboardEvent = require_match.matchesKeyboardEvent;
exports.normalizeHotkey = require_parse.normalizeHotkey;
exports.normalizeKeyName = require_constants.normalizeKeyName;
exports.parseHotkey = require_parse.parseHotkey;
exports.parseKeyboardEvent = require_parse.parseKeyboardEvent;
exports.rawHotkeyToParsedHotkey = require_parse.rawHotkeyToParsedHotkey;
exports.resolveModifier = require_constants.resolveModifier;
exports.validateHotkey = require_validate.validateHotkey;
