// @tanstack/hotkeys/dist/key-state-tracker.cjs
const require_constants = require('./constants.cjs');
let _tanstack_store = require("@tanstack/store");

//#region src/key-state-tracker.ts
/**
* Returns the default state for KeyStateTracker.
*/
function getDefaultKeyStateTrackerState() {
	return {
		heldKeys: [],
		heldCodes: {}
	};
}
/**
* Singleton tracker for currently held keyboard keys.
*
* This class maintains a list of all keys currently being pressed,
* which is useful for:
* - Displaying currently held keys to users
* - Custom shortcut recording for rebinding
* - Complex chord detection
*
* State Management:
* - Uses TanStack Store for reactive state management
* - State can be accessed via `tracker.store.state` when using the class directly
* - When using framework adapters (React), use `useHeldKeys` and `useHeldKeyCodes` hooks for reactive state
*
* @example
* ```ts
* const tracker = KeyStateTracker.getInstance()
*
* // Access state directly
* console.log(tracker.store.state.heldKeys) // ['Control', 'Shift']
*
* // Subscribe to changes with TanStack Store
* const unsubscribe = tracker.store.subscribe(() => {
*   console.log('Currently held:', tracker.store.state.heldKeys)
* })
*
* // Check current state
* console.log(tracker.getHeldKeys()) // ['Control', 'Shift']
* console.log(tracker.isKeyHeld('Control')) // true
*
* // Cleanup
* unsubscribe()
* ```
*/
var KeyStateTracker = class KeyStateTracker {
	static #instance = null;
	#heldKeysSet = /* @__PURE__ */ new Set();
	#heldCodesMap = /* @__PURE__ */ new Map();
	#keydownListener = null;
	#keyupListener = null;
	#blurListener = null;
	constructor() {
		this.store = new _tanstack_store.Store(getDefaultKeyStateTrackerState());
		this.#setupListeners();
	}
	/**
	* Gets the singleton instance of KeyStateTracker.
	*/
	static getInstance() {
		if (!KeyStateTracker.#instance) KeyStateTracker.#instance = new KeyStateTracker();
		return KeyStateTracker.#instance;
	}
	/**
	* Resets the singleton instance. Useful for testing.
	*/
	static resetInstance() {
		if (KeyStateTracker.#instance) {
			KeyStateTracker.#instance.destroy();
			KeyStateTracker.#instance = null;
		}
	}
	/**
	* Sets up the keyboard event listeners.
	*/
	#setupListeners() {
		if (typeof document === "undefined") return;
		this.#keydownListener = (event) => {
			const key = require_constants.normalizeKeyName(event.key);
			if (!this.#heldKeysSet.has(key)) {
				this.#heldKeysSet.add(key);
				this.#heldCodesMap.set(key, event.code);
				this.#syncState();
			}
		};
		this.#keyupListener = (event) => {
			const key = require_constants.normalizeKeyName(event.key);
			if (this.#heldKeysSet.has(key)) {
				this.#heldKeysSet.delete(key);
				this.#heldCodesMap.delete(key);
			}
			if (require_constants.MODIFIER_KEYS.has(key)) {
				for (const heldKey of this.#heldKeysSet) if (!require_constants.MODIFIER_KEYS.has(heldKey)) {
					this.#heldKeysSet.delete(heldKey);
					this.#heldCodesMap.delete(heldKey);
				}
			}
			this.#syncState();
		};
		this.#blurListener = () => {
			if (this.#heldKeysSet.size > 0) {
				this.#heldKeysSet.clear();
				this.#heldCodesMap.clear();
				this.#syncState();
			}
		};
		document.addEventListener("keydown", this.#keydownListener);
		document.addEventListener("keyup", this.#keyupListener);
		window.addEventListener("blur", this.#blurListener);
	}
	/**
	* Syncs the internal Set to the Store state.
	*/
	#syncState() {
		this.store.setState(() => ({
			heldKeys: Array.from(this.#heldKeysSet),
			heldCodes: Object.fromEntries(this.#heldCodesMap)
		}));
	}
	/**
	* Removes the keyboard event listeners.
	*/
	#removeListeners() {
		if (typeof document === "undefined") return;
		if (this.#keydownListener) {
			document.removeEventListener("keydown", this.#keydownListener);
			this.#keydownListener = null;
		}
		if (this.#keyupListener) {
			document.removeEventListener("keyup", this.#keyupListener);
			this.#keyupListener = null;
		}
		if (this.#blurListener) {
			window.removeEventListener("blur", this.#blurListener);
			this.#blurListener = null;
		}
	}
	/**
	* Gets an array of currently held key names.
	*
	* @returns Array of key names currently being pressed
	*/
	getHeldKeys() {
		return this.store.state.heldKeys;
	}
	/**
	* Checks if a specific key is currently being held.
	*
	* @param key - The key name to check (case-insensitive)
	* @returns True if the key is currently held
	*/
	isKeyHeld(key) {
		const normalizedKey = require_constants.normalizeKeyName(key);
		return this.#heldKeysSet.has(normalizedKey);
	}
	/**
	* Checks if any of the given keys are currently held.
	*
	* @param keys - Array of key names to check
	* @returns True if any of the keys are currently held
	*/
	isAnyKeyHeld(keys) {
		return keys.some((key) => this.isKeyHeld(key));
	}
	/**
	* Checks if all of the given keys are currently held.
	*
	* @param keys - Array of key names to check
	* @returns True if all of the keys are currently held
	*/
	areAllKeysHeld(keys) {
		return keys.every((key) => this.isKeyHeld(key));
	}
	/**
	* Destroys the tracker and removes all listeners.
	*/
	destroy() {
		this.#removeListeners();
		this.#heldKeysSet.clear();
		this.#heldCodesMap.clear();
		this.store.setState(() => getDefaultKeyStateTrackerState());
	}
};
/**
* Gets the singleton KeyStateTracker instance.
* Convenience function for accessing the tracker.
*/
function getKeyStateTracker() {
	return KeyStateTracker.getInstance();
}

//#endregion
exports.KeyStateTracker = KeyStateTracker;
exports.getKeyStateTracker = getKeyStateTracker;
