// @tanstack/hotkeys/dist/recorder.js
import { detectPlatform } from "./constants.js";
import { convertToModFormat, hasNonModifierKey, isModifierKey, keyboardEventToHotkey } from "./parse.js";
import { Store } from "@tanstack/store";

//#region src/recorder.ts
/**
* Framework-agnostic class for recording keyboard shortcuts.
*
* This class handles all the complexity of capturing keyboard events,
* converting them to hotkey strings, and handling edge cases like
* Escape to cancel or Backspace/Delete to clear.
*
* State Management:
* - Uses TanStack Store for reactive state management
* - State can be accessed via `recorder.store.state` when using the class directly
* - When using framework adapters (React), use `useStore` hooks for reactive state
*
* @example
* ```ts
* const recorder = new HotkeyRecorder({
*   onRecord: (hotkey) => {
*     console.log('Recorded:', hotkey)
*   },
*   onCancel: () => {
*     console.log('Recording cancelled')
*   },
* })
*
* // Start recording
* recorder.start()
*
* // Access state directly
* console.log(recorder.store.state.isRecording) // true
*
* // Subscribe to changes with TanStack Store
* const unsubscribe = recorder.store.subscribe(() => {
*   console.log('Recording:', recorder.store.state.isRecording)
* })
*
* // Cleanup
* recorder.destroy()
* unsubscribe()
* ```
*/
var HotkeyRecorder = class {
	#keydownHandler = null;
	#options;
	#platform;
	constructor(options) {
		this.store = new Store({
			isRecording: false,
			recordedHotkey: null
		});
		this.#options = options;
		this.#platform = detectPlatform();
	}
	/**
	* Updates the recorder options, including callbacks.
	* This allows framework adapters to sync callback changes without recreating the recorder.
	*/
	setOptions(options) {
		this.#options = {
			...this.#options,
			...options
		};
	}
	/**
	* Start recording a new hotkey.
	*
	* Sets up a keydown event listener that captures keyboard events
	* and converts them to hotkey strings. Recording continues until
	* a valid hotkey is recorded, Escape is pressed, or stop/cancel is called.
	*/
	start() {
		if (this.#keydownHandler) return;
		this.store.setState(() => ({
			isRecording: true,
			recordedHotkey: null
		}));
		const handler = (event) => {
			if (!this.#keydownHandler) return;
			event.preventDefault();
			event.stopPropagation();
			if (event.key === "Escape") {
				this.cancel();
				return;
			}
			if (event.key === "Backspace" || event.key === "Delete") {
				if (!event.ctrlKey && !event.shiftKey && !event.altKey && !event.metaKey) {
					this.#options.onClear?.();
					this.#options.onRecord("");
					this.stop();
					return;
				}
			}
			if (isModifierKey(event)) return;
			const finalHotkey = convertToModFormat(keyboardEventToHotkey(event), this.#platform);
			if (hasNonModifierKey(finalHotkey, this.#platform)) {
				const handlerToRemove = this.#keydownHandler;
				if (handlerToRemove) {
					this.#removeListener(handlerToRemove);
					this.#keydownHandler = null;
				}
				this.store.setState(() => ({
					isRecording: false,
					recordedHotkey: finalHotkey
				}));
				this.#options.onRecord(finalHotkey);
			}
		};
		this.#keydownHandler = handler;
		this.#addListener(handler);
	}
	/**
	* Stop recording (same as cancel, but doesn't call onCancel).
	*
	* Removes the event listener and resets the recording state.
	*/
	stop() {
		if (this.#keydownHandler) {
			this.#removeListener(this.#keydownHandler);
			this.#keydownHandler = null;
		}
		this.store.setState(() => ({
			isRecording: false,
			recordedHotkey: null
		}));
	}
	/**
	* Cancel recording without saving.
	*
	* Removes the event listener, resets the recording state, and calls
	* the onCancel callback if provided.
	*/
	cancel() {
		if (this.#keydownHandler) {
			this.#removeListener(this.#keydownHandler);
			this.#keydownHandler = null;
		}
		this.store.setState(() => ({
			isRecording: false,
			recordedHotkey: null
		}));
		this.#options.onCancel?.();
	}
	/**
	* Adds the keydown event listener to the document.
	*/
	#addListener(handler) {
		if (typeof document === "undefined") return;
		document.addEventListener("keydown", handler, true);
	}
	/**
	* Removes the keydown event listener from the document.
	*/
	#removeListener(handler) {
		if (typeof document === "undefined") return;
		document.removeEventListener("keydown", handler, true);
	}
	/**
	* Clean up event listeners and reset state.
	*
	* Call this when you're done with the recorder to ensure
	* all event listeners are properly removed.
	*/
	destroy() {
		this.stop();
		this.store.setState(() => ({
			isRecording: false,
			recordedHotkey: null
		}));
	}
};

//#endregion
export { HotkeyRecorder };
